
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://localhost:8000/campylobacter-crisprscape/CRISPR_screening.html">
      
      
        <link rel="prev" href="allthebacteria.html">
      
      
        <link rel="next" href="CRISPR_refinement.html">
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.18">
    
    
      
        <title>CRISPR screening - Campylobacter CRISPRscape</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.7e37652d.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
        <script src="https://unpkg.com/iframe-worker/shim"></script>
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#step-by-step-explanation-of-the-crispr-cas-screening-workflow-simplified" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="Campylobacter CRISPRscape" class="md-header__button md-logo" aria-label="Campylobacter CRISPRscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Campylobacter CRISPRscape
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CRISPR screening
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="Campylobacter CRISPRscape" class="md-nav__button md-logo" aria-label="Campylobacter CRISPRscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Campylobacter CRISPRscape
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="manual.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    User manual
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Workflow details
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Workflow details
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Prepare genomes
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Prepare genomes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="prepare_genomes.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Script functions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="allthebacteria.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Metadata
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    CRISPR screening
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="CRISPR_screening.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    CRISPR screening
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-screening-genomes-for-presence-of-crispr-cas" class="md-nav__link">
    <span class="md-ellipsis">
      1. Screening genomes for presence of CRISPR-Cas
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Screening genomes for presence of CRISPR-Cas">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-summarising-the-summaries" class="md-nav__link">
    <span class="md-ellipsis">
      1.1 Summarising the summaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-practical-detail-on-processing-ten-thousands-of-genomes" class="md-nav__link">
    <span class="md-ellipsis">
      1.2 Practical detail on processing ten thousands of genomes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-collect-all-identified-crispr-spacers" class="md-nav__link">
    <span class="md-ellipsis">
      1.3 Collect all identified CRISPR spacers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-files-generated-in-the-process" class="md-nav__link">
    <span class="md-ellipsis">
      Output files generated in the process
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output files generated in the process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="CRISPR_refinement.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    CRISPR refinement
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="clustering_spacers.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    CRISPR clustering
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="spacepharer.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SpacePHARER
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="extra_functions.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Extra functions
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="output_files.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Output files
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="dev_notes.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Development notes
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#1-screening-genomes-for-presence-of-crispr-cas" class="md-nav__link">
    <span class="md-ellipsis">
      1. Screening genomes for presence of CRISPR-Cas
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Screening genomes for presence of CRISPR-Cas">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#11-summarising-the-summaries" class="md-nav__link">
    <span class="md-ellipsis">
      1.1 Summarising the summaries
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#12-practical-detail-on-processing-ten-thousands-of-genomes" class="md-nav__link">
    <span class="md-ellipsis">
      1.2 Practical detail on processing ten thousands of genomes
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#13-collect-all-identified-crispr-spacers" class="md-nav__link">
    <span class="md-ellipsis">
      1.3 Collect all identified CRISPR spacers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output-files-generated-in-the-process" class="md-nav__link">
    <span class="md-ellipsis">
      Output files generated in the process
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Output files generated in the process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#next-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Next steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="step-by-step-explanation-of-the-crispr-cas-screening-workflow-simplified">Step-by-step explanation of the CRISPR-Cas screening workflow (simplified)</h1>
<pre class="mermaid"><code>flowchart LR
    A([Genome]) --&gt;|Screen CRISPR-Cas| B[Tables]
    A([Genome]) --&gt;|Screen CRISPR-Cas| F(FASTA)
    B --&gt;|Parse| C[CSV file]
    B --&gt;|Concatenate| D[Table / batch]
    F(FASTA) --&gt;|Concatenate| G(Spacers / batch)
    C[CSV file] --&gt;|Concatenate| H[CSV / batch]
    H[CSV / batch] --&gt;|"Combine
                        batches"| I[("CRISPR-Cas
                                      database")]</code></pre>
<h2 id="1-screening-genomes-for-presence-of-crispr-cas">1. Screening genomes for presence of CRISPR-Cas <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M10.68 11.74a6 6 0 0 1-7.922-8.982 6 6 0 0 1 8.982 7.922l3.04 3.04a.749.749 0 0 1-.326 1.275.75.75 0 0 1-.734-.215ZM11.5 7a4.499 4.499 0 1 0-8.997 0A4.499 4.499 0 0 0 11.5 7"/></svg></span></h2>
<p>We screen each input genome for the presence of CRISPR-Cas loci using 
<a href="https://github.com/Russel88/CRISPRCasTyper">CCTyper</a>.
For a detailed overview of the method, please see the
<a href="https://doi.org/10.1101/2020.05.15.097824">paper (preprint)</a>.
CCTyper produces tab-separated output tables summarising results of:</p>
<ol>
<li>Complete CRISPR-Cas systems (CRISPR + Cas locus)</li>
<li>Orphan CRISPR arrays (CRISPR spacers and repeats, without cas genes)</li>
<li>Cas operons (Any putative operon of cas genes, with or without CRISPR array)</li>
</ol>
<p>It also saves the CRISPR spacer sequences as <code>.fa</code> file:
each array gets one FASTA file with spacers as separate entries/lines.</p>
<p>This is described in the <code>Snakefile</code> rule <code>crisprcastyper</code>.
The output files have the extension <code>.tab</code>.</p>
<p>CCTyper also provides a visual output for each identified locus, for example:</p>
<p><img alt="Figure 1: CCTyper output" src="figures/cctyper_plot.svg" width="560" /></p>
<h3 id="11-summarising-the-summaries">1.1 Summarising the summaries <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M7.122.392a1.75 1.75 0 0 1 1.756 0l5.003 2.902c.83.481.83 1.68 0 2.162L8.878 8.358a1.75 1.75 0 0 1-1.756 0L2.119 5.456a1.25 1.25 0 0 1 0-2.162ZM8.125 1.69a.25.25 0 0 0-.25 0l-4.63 2.685 4.63 2.685a.25.25 0 0 0 .25 0l4.63-2.685ZM1.601 7.789a.75.75 0 0 1 1.025-.273l5.249 3.044a.25.25 0 0 0 .25 0l5.249-3.044a.75.75 0 0 1 .752 1.298l-5.248 3.044a1.75 1.75 0 0 1-1.756 0L1.874 8.814A.75.75 0 0 1 1.6 7.789Zm0 3.5a.75.75 0 0 1 1.025-.273l5.249 3.044a.25.25 0 0 0 .25 0l5.249-3.044a.75.75 0 0 1 .752 1.298l-5.248 3.044a1.75 1.75 0 0 1-1.756 0l-5.248-3.044a.75.75 0 0 1-.273-1.025"/></svg></span></h3>
<p>To facilitate further processing of the results reported by CCTyper,
we combine the most relevant results from the different <code>.tab</code> files
into one <code>.csv</code> file: <code>CRISPR-Cas.csv</code>. At the same time, the array/
operon names, start and stop positions and DNA sequence orientation
are collected and saved as <code>.bed</code> files.</p>
<p>This corresponds to rule <code>parse_cctyper</code>.
We do this using a custom script: <a href="https://github.com/UtrechtUniversity/campylobacter-crisprscape/blob/main/bin/cctyper_extender.py"><code>bin/cctyper_expender.py</code></a>.</p>
<h3 id="12-practical-detail-on-processing-ten-thousands-of-genomes">1.2 Practical detail on processing ten thousands of genomes <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M9 4.75A.75.75 0 0 1 9.75 4h4a.75.75 0 0 1 .53 1.28l-1.89 1.892c.312.076.604.18.867.319.742.391 1.244 1.063 1.244 2.005 0 .653-.231 1.208-.629 1.627-.386.408-.894.653-1.408.777-1.01.243-2.225.063-3.124-.527a.751.751 0 0 1 .822-1.254c.534.35 1.32.474 1.951.322.306-.073.53-.201.67-.349.129-.136.218-.32.218-.596 0-.308-.123-.509-.444-.678-.373-.197-.98-.318-1.806-.318a.75.75 0 0 1-.53-1.28l1.72-1.72H9.75A.75.75 0 0 1 9 4.75m-3.587 5.763c-.35-.05-.77.113-.983.572a.75.75 0 1 1-1.36-.632c.508-1.094 1.589-1.565 2.558-1.425 1 .145 1.872.945 1.872 2.222 0 1.433-1.088 2.192-1.79 2.681-.308.216-.571.397-.772.573H7a.75.75 0 0 1 0 1.5H3.75a.75.75 0 0 1-.75-.75c0-.69.3-1.211.67-1.61.348-.372.8-.676 1.15-.92.8-.56 1.18-.904 1.18-1.474 0-.473-.267-.69-.587-.737M5.604.089A.75.75 0 0 1 6 .75v4.77h.711a.75.75 0 0 1 0 1.5H3.759a.75.75 0 0 1 0-1.5H4.5V2.15l-.334.223a.75.75 0 0 1-.832-1.248l1.5-1a.75.75 0 0 1 .77-.037Z"/></svg></span></h3>
<p>Since we are working with large numbers of genome files, these have
been separated in batches, each with up to 4,000 genomes. For each batch,
we concatenate the tabular files mentioned above. This is described in
<code>Snakefile</code> rule <code>collect_cctyper</code>.
This step concatenates the results from all <code>.tab</code> files in a batch,
the <code>.csv</code> file generated in <a href="#11-summarising-the-summaries">step 1.1</a>,
and the <code>.fa</code> files containing all separate CRISPR spacer sequences
as generated by CCTyper.</p>
<p>This step uses two custom bash scripts to concatenate tables in a way
that only one header line is used in the concatenated file:
<a href="https://github.com/UtrechtUniversity/campylobacter-crisprscape/blob/main/bin/concatenate_cctyper_output.sh"><code>bin/concatenate_cctyper_output.sh</code></a> and
<a href="https://github.com/UtrechtUniversity/campylobacter-crisprscape/blob/main/bin/concatenate_cctyper_csv.sh"><code>bin/concatenate_cctyper_csv.sh</code></a>.
To concatenate the spacer fasta files, we simply use <code>cat</code>.</p>
<p><em>Note that CCTyper reports putative cas operons.</em>
<em>The <code>concatenate_cctyper_output.sh</code> script also makes a <code>.tab</code> file of</em>
<em>predicted 'true' operons as <code>cas_operons-[batch name].tab</code>.</em></p>
<p>On a technical note, rather than using Snakemake's built-in parallelisation,
we rely on GNU parallel to process genomes in batches. The code is:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>find<span class="w"> </span>-L<span class="w"> </span><span class="o">{</span>input.batch<span class="o">}</span><span class="w"> </span>-mindepth<span class="w"> </span><span class="m">1</span><span class="w"> </span>-maxdepth<span class="w"> </span><span class="m">1</span><span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span>-name<span class="w"> </span><span class="s2">&quot;*.fa&quot;</span><span class="w"> </span>-print0<span class="w"> </span><span class="p">|</span><span class="se">\</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="w"> </span>parallel<span class="w"> </span>-0<span class="w"> </span>--jobs<span class="w"> </span><span class="o">{</span>threads<span class="o">}</span><span class="w"> </span>--retry-failed<span class="w"> </span>--halt<span class="o">=</span><span class="s1">&#39;now,fail=1&#39;</span><span class="se">\</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="w"> </span><span class="s1">&#39;rm -rf &quot;{params.out_dir}{{/.}}&quot; &amp;&amp;\</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="s1"> cctyper -t 1 {{}} &quot;{params.out_dir}{{/.}}&quot;&#39;</span><span class="w"> </span>&gt;<span class="w"> </span><span class="o">{</span>log<span class="o">}</span><span class="w"> </span><span class="m">2</span>&gt;<span class="p">&amp;</span><span class="m">1</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a>touch<span class="w"> </span><span class="o">{</span>output<span class="o">}</span>
</span></code></pre></div>
<p><code>find</code> is used to localise the input files with extension '.fa'.
(<code>-type f</code> matches files only, while the <code>-L</code> option enables the use of
symbolic links.)
These files are then passed to <code>parallel</code>, which executes the codeblock
from <code>rm -rf</code> to <code>cctyper</code>. To prevent funny results with previous analyses,
we remove any previous output directories and then run CCTyper with default
parameters and 1 CPU thread to analyse the given input genome.
All the command-line output (stdout and stderr) is written to the log file
specified in the <code>Snakefile</code>.</p>
<p>For more details on using GNU parallel, see the
<a href="dev_notes.html#use-of-gnu-parallel-over-snakemake">development notes</a>.</p>
<h3 id="13-collect-all-identified-crispr-spacers">1.3 Collect all identified CRISPR spacers <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path d="M5 3.25a.75.75 0 0 1 .75-.75h8.5a.75.75 0 0 1 0 1.5h-8.5A.75.75 0 0 1 5 3.25m0 5a.75.75 0 0 1 .75-.75h8.5a.75.75 0 0 1 0 1.5h-8.5A.75.75 0 0 1 5 8.25m0 5a.75.75 0 0 1 .75-.75h8.5a.75.75 0 0 1 0 1.5h-8.5a.75.75 0 0 1-.75-.75M.924 10.32a.5.5 0 0 1-.851-.525l.001-.001.001-.002.002-.004.007-.011q.147-.217.348-.384c.228-.19.588-.392 1.068-.392.468 0 .858.181 1.126.484.259.294.377.673.377 1.038 0 .987-.686 1.495-1.156 1.845l-.047.035c-.303.225-.522.4-.654.597h1.357a.5.5 0 0 1 0 1H.5a.5.5 0 0 1-.5-.5c0-1.005.692-1.52 1.167-1.875l.035-.025c.531-.396.8-.625.8-1.078a.57.57 0 0 0-.128-.376C1.806 10.068 1.695 10 1.5 10a.66.66 0 0 0-.429.163.8.8 0 0 0-.144.153ZM2.003 2.5V6h.503a.5.5 0 0 1 0 1H.5a.5.5 0 0 1 0-1h.503V3.308l-.28.14a.5.5 0 0 1-.446-.895l1.003-.5a.5.5 0 0 1 .723.447"/></svg></span></h3>
<p>After the concatenation of CCTyper's output files per batch, CRISPR
spacer sequences (as <code>.fa</code> files) for all batches are concatenated
in one batch with the rule <code>concatenate_all_spacers</code>.
This results in a single file <code>all_spacers.fasta</code> for all input genomes together.</p>
<h3 id="output-files-generated-in-the-process">Output files generated in the process <img alt="📁" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@15.1.0/assets/svg/1f4c1.svg" title=":file_folder:" /> <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8zm-4 17H7v-2h3zm0-3H7v-2h3zm0-3H7v-2h3zm4 6h-3v-2h3zm0-3h-3v-2h3zm0-3h-3v-2h3zm-1-4V3.5L18.5 9z"/></svg></span> <span class="twemoji"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8zm-4 17H7v-2h3zm0-3H7v-2h3zm0-3H7v-2h3zm4 6h-3v-2h3zm0-3h-3v-2h3zm0-3h-3v-2h3zm-1-4V3.5L18.5 9z"/></svg></span></h3>
<p>Each step in the process generates a number of output files, which by default
are written to:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>data/
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="w">  </span>tmp/
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="w">    </span>cctyper/<span class="w">            </span><span class="c1"># Here go overall files, such as &#39;all_spacers.fa&#39;</span>
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="w">      </span>batch_<span class="o">[</span>number<span class="o">]</span>/<span class="w">   </span><span class="c1"># Here we get batch-wide files, such as</span>
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="w">                        </span><span class="c1">#  &#39;CRISPR-Cas-batch_[number].csv&#39;.</span>
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="w">        </span><span class="o">[</span>sample_id<span class="o">]</span>/<span class="w">    </span><span class="c1"># Here go the sample-specific files, such as CCTyper&#39;s</span>
</span><span id="__span-1-7"><a id="__codelineno-1-7" name="__codelineno-1-7" href="#__codelineno-1-7"></a><span class="w">                        </span><span class="c1">#  primary output, for example &#39;crisprs_all.tab&#39;</span>
</span></code></pre></div>
<p>For more details on the output files, see <a href="output_files.html">output</a>.</p>
<h4 id="next-steps">Next steps</h4>
<p>&rarr; <a href="clustering_spacers.html">Cluster spacers</a></p>
<p>&rarr; <a href="CRISPR_refinement.html">Refine CRISPRs</a></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": ".", "features": ["content.code.copy", "content.code.select"], "search": "assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.92b07e13.min.js"></script>
      
    
  </body>
</html>