% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  twocolumn,
  landscape]{report}

\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else  
    % xetex/luatex font selection
    \setmainfont[Color=101010]{Open Sans}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{3}
% Make \paragraph and \subparagraph free-standing
\makeatletter
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}{
    \@ifstar
      \xxxParagraphStar
      \xxxParagraphNoStar
  }
  \newcommand{\xxxParagraphStar}[1]{\oldparagraph*{#1}\mbox{}}
  \newcommand{\xxxParagraphNoStar}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}{
    \@ifstar
      \xxxSubParagraphStar
      \xxxSubParagraphNoStar
  }
  \newcommand{\xxxSubParagraphStar}[1]{\oldsubparagraph*{#1}\mbox{}}
  \newcommand{\xxxSubParagraphNoStar}[1]{\oldsubparagraph{#1}\mbox{}}
\fi
\makeatother

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\newenvironment{Shaded}{}{}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.33,0.33}{\textbf{#1}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.42,0.45,0.49}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.36,0.77}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.01,0.18,0.38}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.42,0.45,0.49}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.42,0.45,0.49}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.36,0.77}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{#1}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.36,0.77}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.42,0.45,0.49}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.33,0.33}{\underline{#1}}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{\textbf{#1}}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.36,0.77}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.44,0.26,0.76}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.01,0.18,0.38}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.42,0.45,0.49}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{#1}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.14,0.16,0.18}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.14,0.16,0.18}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.44,0.26,0.76}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.84,0.23,0.29}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.42,0.45,0.49}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.36,0.77}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.01,0.18,0.38}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.01,0.18,0.38}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.89,0.38,0.04}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.01,0.18,0.38}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{1.00,0.33,0.33}{#1}}

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother

\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother

\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}

\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Descriptive statistics},
  pdfauthor={Sam Nooij},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}


\title{Descriptive statistics}
\author{Sam Nooij}
\date{2024-12-03}

\begin{document}
\maketitle

\renewcommand*\contentsname{Table of contents}
{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{2}
\tableofcontents
}

\chapter{\texorpdfstring{CRISPR spacers in
\emph{Campylobacter}}{CRISPR spacers in Campylobacter}}\label{crispr-spacers-in-campylobacter}

\textbf{Descriptive statistics of CRISPR arrays in \emph{Campylobacter}
genomes from
\href{https://allthebacteria.readthedocs.io/}{AllTheBacteria}}

We have downloaded all the \emph{Campylobacter} \emph{jejuni} and
\emph{C. coli} genomes from AllTheBacteria (ATB) to screen for CRISPR
spacers. The current experiment includes all genomes that have been
classified as `Campylobacter\_D jejuni*' or `Campylobacter\_D coli*'
using \href{https://www.nature.com/articles/s41587-024-02412-y}{Sylph}
(which the creators of ATB ran on all genomes), that have been added to
ATB in the incremental update of August 2024. (âˆ—: any species
designation matching these name prefixes were included.) These genomes
have been screened for CRISPR
(\href{https://en.wikipedia.org/wiki/CRISPR}{Clustered Regularly
Interspaced Short Palindromic Repeats}) spacers and \emph{Cas}
(CRISPR-associated) genes using
\href{https://www.biorxiv.org/content/10.1101/2020.05.15.097824v1}{CCTyper}
(version 1.8.0). CCTyper produces tab-separated tables that summarise
the CRISPR spacers, the repeat sequences in between, Cas genes including
statistics and sequences of the spacers.

In this analysis report, we summarise the prevalence and other general
descriptive statistics of the CRISPR spacers. Therefore, we first need
to import the results from CCTyper:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Files were originally processed in batches}
\NormalTok{batches }\OtherTok{\textless{}{-}} \FunctionTok{Sys.glob}\NormalTok{(}\StringTok{"../data/tmp/cctyper/batch\_*"}\NormalTok{)}

\CommentTok{\# Read the tab{-}separated files for each batch and concatenate them}
\NormalTok{read\_multiple\_tables }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(input\_files) \{}
  \FunctionTok{return}\NormalTok{(}
    \FunctionTok{data.frame}\NormalTok{(}\AttributeTok{Filename =}\NormalTok{ input\_files,}
               \FunctionTok{read\_delim}\NormalTok{(input\_files,}
                          \AttributeTok{show\_col\_types =}\NormalTok{ F,}
                          \AttributeTok{trim\_ws =}\NormalTok{ T))}
\NormalTok{  )}
\NormalTok{\}}

\CommentTok{\# Try reading concatenated files, or read separate files from batches}
\CommentTok{\#  and save as one concatenated file}
\NormalTok{read\_or\_create }\OtherTok{\textless{}{-}} \ControlFlowTok{function}\NormalTok{(table\_file) \{}
  \ControlFlowTok{if}\NormalTok{(}\FunctionTok{file.exists}\NormalTok{(table\_file)) \{}
    \FunctionTok{print}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\FunctionTok{basename}\NormalTok{(table\_file), }\StringTok{"exists already, reading from file!"}\NormalTok{))}
    \FunctionTok{return}\NormalTok{(}\FunctionTok{read\_delim}\NormalTok{(table\_file, }\AttributeTok{show\_col\_types =}\NormalTok{ F))}
    
\NormalTok{  \} }\ControlFlowTok{else}\NormalTok{ \{}
    \FunctionTok{print}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}
      \FunctionTok{basename}\NormalTok{(table\_file),}
      \StringTok{"does not yet exist. Concatenating batch tables..."}\NormalTok{))}
    
\NormalTok{    table\_wildcard }\OtherTok{\textless{}{-}} \FunctionTok{gsub}\NormalTok{(}\AttributeTok{x =} \FunctionTok{basename}\NormalTok{(table\_file),}
                           \AttributeTok{pattern =} \StringTok{"concatenated"}\NormalTok{,}
                           \AttributeTok{replace =} \StringTok{"*"}\NormalTok{)}
\NormalTok{    file\_list }\OtherTok{\textless{}{-}} \FunctionTok{Sys.glob}\NormalTok{(}\FunctionTok{paste0}\NormalTok{(batches, }\StringTok{"/"}\NormalTok{, table\_wildcard))}
\NormalTok{    df }\OtherTok{\textless{}{-}} \FunctionTok{do.call}\NormalTok{(}\StringTok{"rbind"}\NormalTok{, }\FunctionTok{lapply}\NormalTok{(file\_list, read\_multiple\_tables))}
    \FunctionTok{write\_delim}\NormalTok{(}\AttributeTok{x =}\NormalTok{ df, }\AttributeTok{file =}\NormalTok{ table\_file, }\AttributeTok{delim =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}
    \FunctionTok{return}\NormalTok{(df)}
\NormalTok{  \}}
\NormalTok{\}}


\NormalTok{CRISPR\_Cas\_df }\OtherTok{\textless{}{-}} \FunctionTok{read\_or\_create}\NormalTok{(}\StringTok{"../data/processed/CRISPR\_Cas{-}concatenated.tab"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "CRISPR_Cas-concatenated.tab exists already, reading from file!"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{crisprs\_all\_df }\OtherTok{\textless{}{-}} \FunctionTok{read\_or\_create}\NormalTok{(}\StringTok{"../data/processed/crisprs\_all{-}concatenated.tab"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "crisprs_all-concatenated.tab exists already, reading from file!"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{crisprs\_near\_cas\_df }\OtherTok{\textless{}{-}} \FunctionTok{read\_or\_create}\NormalTok{(}\StringTok{"../data/processed/crisprs\_near\_cas{-}concatenated.tab"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{Contig) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{separate\_wider\_delim}\NormalTok{(}\AttributeTok{cols =}\NormalTok{ CRISPR, }\AttributeTok{delim =} \StringTok{"."}\NormalTok{,}
                       \AttributeTok{names =} \FunctionTok{c}\NormalTok{(}\StringTok{"sample\_accession"}\NormalTok{, }\StringTok{"Contig"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{separate\_wider\_delim}\NormalTok{(}\AttributeTok{cols =}\NormalTok{ Contig, }\AttributeTok{delim =} \StringTok{"\_"}\NormalTok{,}
                       \AttributeTok{names =} \FunctionTok{c}\NormalTok{(}\StringTok{"Contig"}\NormalTok{, }\StringTok{"CRISPR\_number"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "crisprs_near_cas-concatenated.tab exists already, reading from file!"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{crisprs\_orphan\_df }\OtherTok{\textless{}{-}} \FunctionTok{read\_or\_create}\NormalTok{(}\StringTok{"../data/processed/crisprs\_orphan{-}concatenated.tab"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{Contig) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{separate\_wider\_delim}\NormalTok{(}\AttributeTok{cols =}\NormalTok{ CRISPR, }\AttributeTok{delim =} \StringTok{"."}\NormalTok{,}
                       \AttributeTok{names =} \FunctionTok{c}\NormalTok{(}\StringTok{"sample\_accession"}\NormalTok{, }\StringTok{"Contig"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{separate\_wider\_delim}\NormalTok{(}\AttributeTok{cols =}\NormalTok{ Contig, }\AttributeTok{delim =} \StringTok{"\_"}\NormalTok{,}
                       \AttributeTok{names =} \FunctionTok{c}\NormalTok{(}\StringTok{"Contig"}\NormalTok{, }\StringTok{"CRISPR\_number"}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "crisprs_orphan-concatenated.tab exists already, reading from file!"
\end{verbatim}

The output of CCTyper only contains sample accession IDs that can be
used to identify samples. Import a minimal set of ENA metadata and
species identifications by Sylph:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{campylobacter\_metadata\_file }\OtherTok{\textless{}{-}} \StringTok{"../data/processed/Campylobacter\_ATB\_metadata.tsv"}

\ControlFlowTok{if}\NormalTok{(}\FunctionTok{file.exists}\NormalTok{(campylobacter\_metadata\_file)) \{}
  \FunctionTok{message}\NormalTok{(}\StringTok{"Reading Campylobacter metadata file!"}\NormalTok{)}
\NormalTok{  campylobacter\_metadata }\OtherTok{\textless{}{-}} \FunctionTok{read\_delim}\NormalTok{(}
\NormalTok{    campylobacter\_metadata\_file,}
    \AttributeTok{show\_col\_types =}\NormalTok{ F)}
  
\NormalTok{\} }\ControlFlowTok{else}\NormalTok{ \{}
  \FunctionTok{message}\NormalTok{(}\StringTok{"Creating Campylobacter metadata file..."}\NormalTok{)}
\NormalTok{  total\_campylobacter\_metadata }\OtherTok{\textless{}{-}} \FunctionTok{read\_delim}\NormalTok{(}
    \StringTok{"../data/ATB/ena\_metadata.20240801.selection{-}only\_Campylobacter.tsv.gz"}\NormalTok{,}
    \AttributeTok{delim =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{,}
    \AttributeTok{col\_names =}\NormalTok{ F,}
    \AttributeTok{show\_col\_types =}\NormalTok{ F)}
  
\NormalTok{  older\_campylobacter\_metadata }\OtherTok{\textless{}{-}} \FunctionTok{read\_delim}\NormalTok{(}
    \StringTok{"../data/ATB/ena\_metadata.0.2.20240606.selection{-}only\_Campylobacter.tsv.gz"}\NormalTok{,}
    \AttributeTok{delim =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{,}
    \AttributeTok{col\_names =}\NormalTok{ F,}
    \AttributeTok{show\_col\_types =}\NormalTok{ F)}
  
  \CommentTok{\# Remove the older entries to keep only the Campylobacter genomes}
  \CommentTok{\# from the incremental update}
\NormalTok{  campylobacter\_metadata }\OtherTok{\textless{}{-}}\NormalTok{ total\_campylobacter\_metadata }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{filter}\NormalTok{(}\SpecialCharTok{!}\NormalTok{ X2 }\SpecialCharTok{\%in\%}\NormalTok{ older\_campylobacter\_metadata}\SpecialCharTok{$}\NormalTok{X1)}
  
  \CommentTok{\# Take the column names from the table before grepping}
  \FunctionTok{colnames}\NormalTok{(campylobacter\_metadata) }\OtherTok{\textless{}{-}} \FunctionTok{read\_delim}\NormalTok{(}
    \AttributeTok{file =} \StringTok{"../data/ATB/ena\_metadata.20240801.selection.tsv.gz"}\NormalTok{,}
    \AttributeTok{n\_max =} \DecValTok{1}\NormalTok{,}
    \AttributeTok{delim =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{,}
    \AttributeTok{show\_col\_types =}\NormalTok{ F}
\NormalTok{  ) }\SpecialCharTok{\%\textgreater{}\%} \FunctionTok{colnames}\NormalTok{()}
  
  \CommentTok{\# Remove columns that contain only NA values}
\NormalTok{  campylobacter\_metadata }\OtherTok{\textless{}{-}}\NormalTok{ campylobacter\_metadata }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{select\_if}\NormalTok{(}\SpecialCharTok{\textasciitilde{}}\FunctionTok{sum}\NormalTok{(}\SpecialCharTok{!}\FunctionTok{is.na}\NormalTok{(.)) }\SpecialCharTok{\textgreater{}} \DecValTok{0}\NormalTok{)}
  
\NormalTok{  simplified\_species }\OtherTok{\textless{}{-}} \FunctionTok{read\_delim}\NormalTok{(}
    \AttributeTok{file =} \StringTok{"../data/ATB/sylph.tsv.gz"}\NormalTok{,}
    \AttributeTok{delim =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{,}
    \AttributeTok{show\_col\_types =}\NormalTok{ F) }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{select}\NormalTok{(Sample, Species)}
  
\NormalTok{  campylobacter\_metadata }\OtherTok{\textless{}{-}} \FunctionTok{left\_join}\NormalTok{(}
    \AttributeTok{x =}\NormalTok{ campylobacter\_metadata,}
    \AttributeTok{y =}\NormalTok{ simplified\_species,}
    \AttributeTok{by =} \FunctionTok{c}\NormalTok{(}\StringTok{"sample\_accession"} \OtherTok{=} \StringTok{"Sample"}\NormalTok{)}
\NormalTok{  )}
  
  \CommentTok{\# Remove all genomes whose taxonomic annotation (species) does}
  \CommentTok{\# not start with \textquotesingle{}Campylobacter\textquotesingle{}}
\NormalTok{  campylobacter\_metadata }\OtherTok{\textless{}{-}}\NormalTok{ campylobacter\_metadata }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{filter}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\StringTok{"Campylobacter\_D jejuni|Campylobacter\_D coli"}\NormalTok{, Species))}
  
  \FunctionTok{write\_delim}\NormalTok{(}\AttributeTok{x =}\NormalTok{ campylobacter\_metadata,}
              \AttributeTok{file =}\NormalTok{ campylobacter\_metadata\_file,}
              \AttributeTok{delim =} \StringTok{"}\SpecialCharTok{\textbackslash{}t}\StringTok{"}\NormalTok{)}
\NormalTok{\}}

\NormalTok{campylobacter\_species\_numbers }\OtherTok{\textless{}{-}}\NormalTok{ campylobacter\_metadata }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{species =} \FunctionTok{if\_else}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\StringTok{"jejuni"}\NormalTok{, Species), }\StringTok{"jejuni"}\NormalTok{, }\StringTok{"coli"}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{count}\NormalTok{(species)}

\CommentTok{\# Alternative visualisation method: stacked bar chart}
\NormalTok{stacked\_species\_plot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}\AttributeTok{data =}\NormalTok{ campylobacter\_species\_numbers,}
                             \FunctionTok{aes}\NormalTok{(}\AttributeTok{x =} \DecValTok{0}\NormalTok{, }\AttributeTok{y =}\NormalTok{ n, }\AttributeTok{fill =}\NormalTok{ species)) }\SpecialCharTok{+}
  \FunctionTok{geom\_bar}\NormalTok{(}\AttributeTok{stat =} \StringTok{"identity"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_pubr}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{ggtitle}\NormalTok{(}\FunctionTok{paste}\NormalTok{(}\StringTok{"Total genomes:"}\NormalTok{,}
                \FunctionTok{format}\NormalTok{(}\FunctionTok{sum}\NormalTok{(campylobacter\_species\_numbers}\SpecialCharTok{$}\NormalTok{n),}
                       \AttributeTok{big.mark =} \StringTok{","}\NormalTok{))) }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.title =} \FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.ticks.y =} \FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{axis.text.y =} \FunctionTok{element\_blank}\NormalTok{(),}
        \AttributeTok{legend.position =} \StringTok{"right"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{scale\_fill\_manual}\NormalTok{(}\AttributeTok{name =} \StringTok{"Species"}\NormalTok{,}
                    \AttributeTok{values =} \FunctionTok{c}\NormalTok{(}\StringTok{"\#895d45"}\NormalTok{, }\StringTok{"\#c6903c"}\NormalTok{)) }\SpecialCharTok{+}
  \FunctionTok{coord\_flip}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{scale\_y\_continuous}\NormalTok{(}\AttributeTok{expand =} \FunctionTok{expansion}\NormalTok{(}\AttributeTok{mult =} \FunctionTok{c}\NormalTok{(}\DecValTok{0}\NormalTok{, }\FloatTok{0.01}\NormalTok{))) }\SpecialCharTok{+}
  \FunctionTok{geom\_text}\NormalTok{(}\FunctionTok{aes}\NormalTok{(}\AttributeTok{label =}\NormalTok{ n }\SpecialCharTok{\%\textgreater{}\%}
                  \FunctionTok{format}\NormalTok{(}\AttributeTok{big.mark =} \StringTok{","}\NormalTok{)),}
            \AttributeTok{position =} \FunctionTok{position\_stack}\NormalTok{(}\AttributeTok{vjust =} \FloatTok{0.5}\NormalTok{),}
            \AttributeTok{size =} \FloatTok{4.5}\NormalTok{,}
            \AttributeTok{color =} \StringTok{"black"}\NormalTok{)}

\NormalTok{stacked\_species\_plot}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

\centering{

\includegraphics{Descriptive_statistics_files/figure-pdf/fig-number_of_species-1.pdf}

}

\caption{\label{fig-number_of_species}Number of species present in
dataset}

\end{figure}%

\chapter{Spacers per genome}\label{spacers-per-genome}

We have identified CRISPR spacer in all \emph{Campylobacter jejuni} and
\emph{C. coli} genomes using
\href{https://github.com/Russel88/CRISPRCasTyper}{CCTyper}. CCTyper
separately annotates CRISPR spacers near \emph{Cas} genes and separate
CRISPR arrays, called `orphans'. To inventory the prevalence and size of
both types of CRISPR arrays, show for each genome the number of CRISPR
spacers as scatterplot (Figure~\ref{fig-crisprs_per_genome}).

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{annotated\_crispr\_number\_df }\OtherTok{\textless{}{-}} \FunctionTok{left\_join}\NormalTok{(}
  \AttributeTok{x =}\NormalTok{ campylobacter\_metadata }\SpecialCharTok{\%\textgreater{}\%}
    \FunctionTok{select}\NormalTok{(sample\_accession, country, host, isolation\_source,}
\NormalTok{           scientific\_name, tax\_id, Species, description),}
  \AttributeTok{y =} \FunctionTok{full\_join}\NormalTok{(}
    \AttributeTok{x =}\NormalTok{ crisprs\_near\_cas\_df }\SpecialCharTok{\%\textgreater{}\%}
      \FunctionTok{group\_by}\NormalTok{(sample\_accession) }\SpecialCharTok{\%\textgreater{}\%}
      \FunctionTok{count}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%}
      \FunctionTok{rename}\NormalTok{(}\StringTok{"CRISPRs\_near\_Cas"} \OtherTok{=} \StringTok{"n"}\NormalTok{),}
    \AttributeTok{y =}\NormalTok{ crisprs\_orphan\_df }\SpecialCharTok{\%\textgreater{}\%}
      \FunctionTok{group\_by}\NormalTok{(sample\_accession) }\SpecialCharTok{\%\textgreater{}\%}
      \FunctionTok{count}\NormalTok{() }\SpecialCharTok{\%\textgreater{}\%}
      \FunctionTok{rename}\NormalTok{(}\StringTok{"Orphan\_CRISPRs"} \OtherTok{=} \StringTok{"n"}\NormalTok{),}
    \AttributeTok{by =} \StringTok{"sample\_accession"}
\NormalTok{  )}
\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}
    \FunctionTok{across}\NormalTok{(CRISPRs\_near\_Cas}\SpecialCharTok{:}\NormalTok{Orphan\_CRISPRs,}
           \SpecialCharTok{\textasciitilde{}}\FunctionTok{replace\_na}\NormalTok{(., }\DecValTok{0}\NormalTok{))}
\NormalTok{  )}

\NormalTok{number\_of\_crisprs\_scatterplot }\OtherTok{\textless{}{-}} \FunctionTok{ggplot}\NormalTok{(}
  \AttributeTok{data =}\NormalTok{ annotated\_crispr\_number\_df,}
  \AttributeTok{mapping =} \FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ CRISPRs\_near\_Cas,}
                \AttributeTok{y =}\NormalTok{ Orphan\_CRISPRs)}
\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{geom\_point}\NormalTok{(}\AttributeTok{position =} \FunctionTok{position\_jitter}\NormalTok{(),}
             \AttributeTok{alpha =} \FloatTok{0.3}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_pubr}\NormalTok{()}

\NormalTok{crispr\_near\_cas\_density }\OtherTok{=} \FunctionTok{ggplot}\NormalTok{(}
  \AttributeTok{data =}\NormalTok{ annotated\_crispr\_number\_df,}
  \AttributeTok{mapping =} \FunctionTok{aes}\NormalTok{(}\AttributeTok{x =}\NormalTok{ CRISPRs\_near\_Cas)) }\SpecialCharTok{+}
  \FunctionTok{geom\_density}\NormalTok{(}\AttributeTok{fill =} \StringTok{"grey20"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_pubr}\NormalTok{() }\SpecialCharTok{+}
  \FunctionTok{theme}\NormalTok{(}\AttributeTok{axis.title.x =} \FunctionTok{element\_blank}\NormalTok{())}

\NormalTok{orphan\_crispr\_density }\OtherTok{=} \FunctionTok{ggplot}\NormalTok{(}
  \AttributeTok{data =}\NormalTok{ annotated\_crispr\_number\_df,}
  \AttributeTok{mapping =} \FunctionTok{aes}\NormalTok{(}\AttributeTok{y =}\NormalTok{ Orphan\_CRISPRs)) }\SpecialCharTok{+}
  \FunctionTok{geom\_density}\NormalTok{(}\AttributeTok{fill =} \StringTok{"grey20"}\NormalTok{) }\SpecialCharTok{+}
  \FunctionTok{theme\_pubr}\NormalTok{()}

\NormalTok{design }\OtherTok{\textless{}{-}} \StringTok{"}
\StringTok{1111\#}
\StringTok{22223}
\StringTok{22223}
\StringTok{"}

\NormalTok{spacers\_per\_genome\_figure }\OtherTok{\textless{}{-}}\NormalTok{ crispr\_near\_cas\_density }\SpecialCharTok{+}
\NormalTok{  number\_of\_crisprs\_scatterplot }\SpecialCharTok{+}
\NormalTok{  orphan\_crispr\_density }\SpecialCharTok{+}
  \FunctionTok{plot\_layout}\NormalTok{(}\AttributeTok{design =}\NormalTok{ design,}
              \AttributeTok{axes =} \StringTok{"collect"}\NormalTok{)}

\FunctionTok{print}\NormalTok{(spacers\_per\_genome\_figure)}
\end{Highlighting}
\end{Shaded}

\begin{figure}[H]

\centering{

\includegraphics{Descriptive_statistics_files/figure-pdf/fig-crisprs_per_genome-1.pdf}

}

\caption{\label{fig-crisprs_per_genome}Number of CRISPR spacers per
genome}

\end{figure}%

From this figure, it appears that by far the most genomes have no CRISPR
spacers at all, and otherwise they mostly have one CRISPR spacer near
\emph{Cas} genes. At most, CCTyper has found 3 CRISPRs near \emph{Cas}
genes and up to 4 orphan CRISPR spacers. Let's see this distribution in
numbers and separate them between \emph{C. jejuni} and \emph{coli}.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{total\_genomes }\OtherTok{=} \FunctionTok{nrow}\NormalTok{(annotated\_crispr\_number\_df)}

\NormalTok{annotated\_crispr\_number\_df }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{count}\NormalTok{(CRISPRs\_near\_Cas, }\AttributeTok{name =} \StringTok{"Genomes"}\NormalTok{) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{Percentage =} \FunctionTok{round}\NormalTok{(Genomes }\SpecialCharTok{/}\NormalTok{ total\_genomes }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{2}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 4 x 3
  CRISPRs_near_Cas Genomes Percentage
             <int>   <int>      <dbl>
1                0    8443      43.4 
2                1   10744      55.2 
3                2     256       1.32
4                3       2       0.01
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(}\StringTok{"CRISPRs in Campylobacter jejuni"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "CRISPRs in Campylobacter jejuni"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{annotated\_crispr\_number\_df }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\AttributeTok{pattern =} \StringTok{"Campylobacter\_D jejuni"}\NormalTok{, Species)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{group\_by}\NormalTok{(CRISPRs\_near\_Cas) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarise}\NormalTok{(}\AttributeTok{Genomes =} \FunctionTok{length}\NormalTok{(CRISPRs\_near\_Cas)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{total\_genomes =} \FunctionTok{sum}\NormalTok{(Genomes),}
         \AttributeTok{Percentage =} \FunctionTok{round}\NormalTok{(Genomes }\SpecialCharTok{/}\NormalTok{ total\_genomes }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{2}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{total\_genomes)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 3 x 3
  CRISPRs_near_Cas Genomes Percentage
             <int>   <int>      <dbl>
1                0    4493      32.4 
2                1    9149      66.0 
3                2     231       1.67
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# The number of other jejuni subspecies(?) is too low to do}
\CommentTok{\# separate analyses, so just put them all together}

\FunctionTok{print}\NormalTok{(}\StringTok{"Orphan CRISPRs in C. jejuni"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "Orphan CRISPRs in C. jejuni"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{annotated\_crispr\_number\_df }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\AttributeTok{pattern =} \StringTok{"Campylobacter\_D jejuni"}\NormalTok{, Species)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{group\_by}\NormalTok{(Orphan\_CRISPRs) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarise}\NormalTok{(}\AttributeTok{Genomes =} \FunctionTok{length}\NormalTok{(Orphan\_CRISPRs)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{total\_genomes =} \FunctionTok{sum}\NormalTok{(Genomes),}
         \AttributeTok{Percentage =} \FunctionTok{round}\NormalTok{(Genomes }\SpecialCharTok{/}\NormalTok{ total\_genomes }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{2}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{total\_genomes)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 3 x 3
  Orphan_CRISPRs Genomes Percentage
           <int>   <int>      <dbl>
1              0   13319      96.0 
2              1     539       3.89
3              2      15       0.11
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(}\StringTok{"CRISPRs in Campylobacter coli"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "CRISPRs in Campylobacter coli"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{annotated\_crispr\_number\_df }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\AttributeTok{pattern =} \StringTok{"Campylobacter\_D coli"}\NormalTok{, Species)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{group\_by}\NormalTok{(CRISPRs\_near\_Cas) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarise}\NormalTok{(}\AttributeTok{Genomes =} \FunctionTok{length}\NormalTok{(CRISPRs\_near\_Cas)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{total\_genomes =} \FunctionTok{sum}\NormalTok{(Genomes),}
         \AttributeTok{Percentage =} \FunctionTok{round}\NormalTok{(Genomes }\SpecialCharTok{/}\NormalTok{ total\_genomes }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{2}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{total\_genomes)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 4 x 3
  CRISPRs_near_Cas Genomes Percentage
             <int>   <int>      <dbl>
1                0    3950      70.9 
2                1    1595      28.6 
3                2      25       0.45
4                3       2       0.04
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{print}\NormalTok{(}\StringTok{"Orphan CRISPRs in C. coli"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
[1] "Orphan CRISPRs in C. coli"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{annotated\_crispr\_number\_df }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{filter}\NormalTok{(}\FunctionTok{grepl}\NormalTok{(}\AttributeTok{pattern =} \StringTok{"Campylobacter\_D coli"}\NormalTok{, Species)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{group\_by}\NormalTok{(Orphan\_CRISPRs) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{summarise}\NormalTok{(}\AttributeTok{Genomes =} \FunctionTok{length}\NormalTok{(Orphan\_CRISPRs)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{mutate}\NormalTok{(}\AttributeTok{total\_genomes =} \FunctionTok{sum}\NormalTok{(Genomes),}
         \AttributeTok{Percentage =} \FunctionTok{round}\NormalTok{(Genomes }\SpecialCharTok{/}\NormalTok{ total\_genomes }\SpecialCharTok{*} \DecValTok{100}\NormalTok{, }\DecValTok{2}\NormalTok{)) }\SpecialCharTok{\%\textgreater{}\%}
  \FunctionTok{select}\NormalTok{(}\SpecialCharTok{{-}}\NormalTok{total\_genomes)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
# A tibble: 5 x 3
  Orphan_CRISPRs Genomes Percentage
           <int>   <int>      <dbl>
1              0    5390      96.7 
2              1     170       3.05
3              2      10       0.18
4              3       1       0.02
5              4       1       0.02
\end{verbatim}




\end{document}
