<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Sam Nooij">
<meta name="dcterms.date" content="2024-12-03">

<title>Descriptive statistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Descriptive_statistics_files/libs/clipboard/clipboard.min.js"></script>
<script src="Descriptive_statistics_files/libs/quarto-html/quarto.js"></script>
<script src="Descriptive_statistics_files/libs/quarto-html/popper.min.js"></script>
<script src="Descriptive_statistics_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Descriptive_statistics_files/libs/quarto-html/anchor.min.js"></script>
<link href="Descriptive_statistics_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Descriptive_statistics_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Descriptive_statistics_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Descriptive_statistics_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Descriptive_statistics_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="Descriptive_statistics_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="Descriptive_statistics_files/libs/pagedtable-1.1/js/pagedtable.js"></script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#crispr-spacers-in-campylobacter" id="toc-crispr-spacers-in-campylobacter" class="nav-link active" data-scroll-target="#crispr-spacers-in-campylobacter">CRISPR spacers in <em>Campylobacter</em></a></li>
  <li><a href="#crispr-arrays-per-genome" id="toc-crispr-arrays-per-genome" class="nav-link" data-scroll-target="#crispr-arrays-per-genome">CRISPR arrays per genome</a>
  <ul class="collapse">
  <li><a href="#arrays-per-genome-and-per-species" id="toc-arrays-per-genome-and-per-species" class="nav-link" data-scroll-target="#arrays-per-genome-and-per-species">Arrays per genome and per species</a></li>
  <li><a href="#type-of-crispr-arrays" id="toc-type-of-crispr-arrays" class="nav-link" data-scroll-target="#type-of-crispr-arrays">Type of CRISPR arrays</a></li>
  </ul></li>
  <li><a href="#crispr-spacers-per-genome" id="toc-crispr-spacers-per-genome" class="nav-link" data-scroll-target="#crispr-spacers-per-genome">CRISPR spacers per genome</a>
  <ul class="collapse">
  <li><a href="#spacers-per-species" id="toc-spacers-per-species" class="nav-link" data-scroll-target="#spacers-per-species">Spacers per species</a></li>
  </ul></li>
  <li><a href="#length-of-spacers-and-repeats" id="toc-length-of-spacers-and-repeats" class="nav-link" data-scroll-target="#length-of-spacers-and-repeats">Length of spacers and repeats</a></li>
  <li><a href="#repeat-sequence-summary" id="toc-repeat-sequence-summary" class="nav-link" data-scroll-target="#repeat-sequence-summary">Repeat sequence summary</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="Descriptive_statistics.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Descriptive statistics</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliation</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Sam Nooij <a href="mailto:s.nooij [at] uu.nl" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-5892-5637" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Utrecht University
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 3, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="crispr-spacers-in-campylobacter" class="level1">
<h1>CRISPR spacers in <em>Campylobacter</em></h1>
<p><strong>Descriptive statistics of CRISPR arrays in <em>Campylobacter</em> genomes from <a href="https://allthebacteria.readthedocs.io/" title="AllTheBacteria documentation">AllTheBacteria</a></strong></p>
<p>We have downloaded all the <em>Campylobacter</em> <em>jejuni</em> and <em>C. coli</em> genomes from AllTheBacteria (ATB) to screen for CRISPR spacers. The current experiment includes all genomes that have been classified as ‘Campylobacter_D jejuni*’ or ‘Campylobacter_D coli*’ using <a href="https://www.nature.com/articles/s41587-024-02412-y" title="Shaw &amp; Yu, 2024. Nature Biotechnology">Sylph</a> (which the creators of ATB ran on all genomes), that have been added to ATB in the incremental update of August 2024 (<a href="#fig-number_of_species" class="quarto-xref">Figure&nbsp;1</a>). (∗: any species designation matching these name prefixes were included.) These genomes have been screened for CRISPR (<a href="https://en.wikipedia.org/wiki/CRISPR" title="CRISPR wiki">Clustered Regularly Interspaced Short Palindromic Repeats</a>) spacers and <em>Cas</em> (CRISPR-associated) genes using <a href="https://www.biorxiv.org/content/10.1101/2020.05.15.097824v1" title="Russel et al., 2020. bioRxiv">CCTyper</a> (version 1.8.0). CCTyper produces tab-separated tables that summarise the CRISPRs, including number of contigs containing CRISPR arrays, number of spacers per array, number of repeat sequences, consensus repeat sequence, and presence of Cas genes. From these results, we can view:</p>
<ol type="1">
<li><p>The <a href="#crispr-arrays-per-genome">number of CRISPR arrays</a> detected per genome (<a href="#fig-crisprs_per_genome" class="quarto-xref">Figure&nbsp;2</a>)<br>
(Taking into account the presence of <em>Cas</em> genes)</p></li>
<li><p><a href="#arrays-per-genome-and-per-species">Prevalence of CRISPR arrays</a> overall and per species (<a href="#fig-crisprs_per_genome_donut" class="quarto-xref">Figure&nbsp;3</a>)</p></li>
<li><p><a href="#type-of-crispr-arrays">Type of CRISPR arrays</a> (and prevalence)</p></li>
<li><p>Number of CRISPR <a href="#0">spacers per genome</a> (<a href="#fig-spacers_per_genome" class="quarto-xref">Figure&nbsp;4</a>) and <a href="#0">per species</a> (<a href="#fig-total_spacers_per_species" class="quarto-xref">Figure&nbsp;5</a>)</p></li>
<li><p><a href="#length-of-spacers-and-repeats">Length of the CRISPR</a> spacers and repeat sequences (<a href="#fig-repeat_and_spacer_length" class="quarto-xref">Figure&nbsp;6</a>)</p></li>
<li><p><a href="#repeat-sequence-summary">Repeat consensus sequence diversity</a></p></li>
<li><p>Extra: estimated genome assembly quality (CheckM completeness and contamination)</p></li>
</ol>
<p><em>Note: see the relevant <a href="https://github.com/Russel88/CRISPRCasTyper?tab=readme-ov-file#output-" title="CCTyper output files documentation">CCTyper documentation</a> of output files.</em></p>
<p>In this analysis report, we summarise the prevalence and other general descriptive statistics of the CRISPR spacers. Therefore, we first need to import the results from CCTyper:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Files were originally processed in batches</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>batches <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="fu">here</span>(<span class="st">"data/tmp/cctyper/batch_*"</span>))</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the tab-separated files for each batch and concatenate them</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>read_multiple_tables <span class="ot">&lt;-</span> <span class="cf">function</span>(input_files) {</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(<span class="at">Filename =</span> input_files,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>               <span class="fu">read_delim</span>(input_files,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">show_col_types =</span> F,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">trim_ws =</span> T))</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Try reading concatenated files, or read separate files from batches</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co">#  and save as one concatenated file</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>read_or_create <span class="ot">&lt;-</span> <span class="cf">function</span>(table_file) {</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">file.exists</span>(table_file)) {</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">paste</span>(<span class="fu">basename</span>(table_file), <span class="st">"exists already, reading from file!"</span>))</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">read_delim</span>(table_file, <span class="at">show_col_types =</span> F))</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">message</span>(<span class="fu">paste</span>(</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>      <span class="fu">basename</span>(table_file),</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>      <span class="st">"does not yet exist. Concatenating batch tables..."</span>))</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>    table_wildcard <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">x =</span> <span class="fu">basename</span>(table_file),</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>                           <span class="at">pattern =</span> <span class="st">"concatenated"</span>,</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>                           <span class="at">replace =</span> <span class="st">"*"</span>)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>    file_list <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="fu">paste0</span>(batches, <span class="st">"/"</span>, table_wildcard))</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(file_list, read_multiple_tables))</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">write_delim</span>(<span class="at">x =</span> df, <span class="at">file =</span> table_file, <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(df)</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>CRISPR_Cas_df <span class="ot">&lt;-</span> <span class="fu">read_or_create</span>(</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">here</span>(<span class="st">"data/processed/CRISPR_Cas-concatenated.tab"</span>)</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>crisprs_all_df <span class="ot">&lt;-</span> <span class="fu">read_or_create</span>(</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>  <span class="fu">here</span>(<span class="st">"data/processed/crisprs_all-concatenated.tab"</span>)</span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>crisprs_near_cas_df <span class="ot">&lt;-</span> <span class="fu">read_or_create</span>(</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">here</span>(<span class="st">"data/processed/crisprs_near_cas-concatenated.tab"</span>)</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>Contig) <span class="sc">%&gt;%</span></span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate_wider_delim</span>(<span class="at">cols =</span> CRISPR, <span class="at">delim =</span> <span class="st">"."</span>,</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>                       <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"sample_accession"</span>, <span class="st">"Contig"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate_wider_delim</span>(<span class="at">cols =</span> Contig, <span class="at">delim =</span> <span class="st">"_"</span>,</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>                       <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"Contig"</span>, <span class="st">"CRISPR_number"</span>))</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>crisprs_orphan_df <span class="ot">&lt;-</span> <span class="fu">read_or_create</span>(</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">here</span>(<span class="st">"data/processed/crisprs_orphan-concatenated.tab"</span>)</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>Contig) <span class="sc">%&gt;%</span></span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate_wider_delim</span>(<span class="at">cols =</span> CRISPR, <span class="at">delim =</span> <span class="st">"."</span>,</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a>                       <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"sample_accession"</span>, <span class="st">"Contig"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate_wider_delim</span>(<span class="at">cols =</span> Contig, <span class="at">delim =</span> <span class="st">"_"</span>,</span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a>                       <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"Contig"</span>, <span class="st">"CRISPR_number"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The output of CCTyper only contains sample accession IDs that can be used to identify samples. Import a minimal set of ENA metadata and species identifications by Sylph:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>campylobacter_metadata_file <span class="ot">&lt;-</span> <span class="fu">here</span>(<span class="st">"data/processed/Campylobacter_ATB_metadata.tsv"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">file.exists</span>(campylobacter_metadata_file)) {</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">"Reading Campylobacter metadata file!"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  campylobacter_metadata <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    campylobacter_metadata_file,</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> F)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">"Creating Campylobacter metadata file..."</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  total_campylobacter_metadata <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(<span class="st">"data/ATB/ena_metadata.20240801.selection-only_Campylobacter.tsv.gz"</span>),</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> F,</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> F)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  older_campylobacter_metadata <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">here</span>(<span class="st">"data/ATB/ena_metadata.0.2.20240606.selection-only_Campylobacter.tsv.gz"</span>),</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> F,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> F)</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove the older entries to keep only the Campylobacter genomes</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># from the incremental update</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  campylobacter_metadata <span class="ot">&lt;-</span> total_campylobacter_metadata <span class="sc">%&gt;%</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="sc">!</span> X2 <span class="sc">%in%</span> older_campylobacter_metadata<span class="sc">$</span>X1)</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Take the column names from the table before grepping</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(campylobacter_metadata) <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"data/ATB/ena_metadata.20240801.selection.tsv.gz"</span>),</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_max =</span> <span class="dv">1</span>,</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>    <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> F</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span> <span class="fu">colnames</span>()</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove columns that contain only NA values</span></span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>  campylobacter_metadata <span class="ot">&lt;-</span> campylobacter_metadata <span class="sc">%&gt;%</span></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select_if</span>(<span class="sc">~</span><span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(.)) <span class="sc">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>  simplified_species <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a>    <span class="at">file =</span> <span class="fu">here</span>(<span class="st">"data/ATB/sylph.tsv.gz"</span>),</span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>    <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    <span class="at">show_col_types =</span> F) <span class="sc">%&gt;%</span></span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(Sample, Species)</span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>  campylobacter_metadata <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> campylobacter_metadata,</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> simplified_species,</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"sample_accession"</span> <span class="ot">=</span> <span class="st">"Sample"</span>)</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove all genomes whose taxonomic annotation (species) does</span></span>
<span id="cb2-53"><a href="#cb2-53" aria-hidden="true" tabindex="-1"></a>  <span class="co"># not start with 'Campylobacter'</span></span>
<span id="cb2-54"><a href="#cb2-54" aria-hidden="true" tabindex="-1"></a>  campylobacter_metadata <span class="ot">&lt;-</span> campylobacter_metadata <span class="sc">%&gt;%</span></span>
<span id="cb2-55"><a href="#cb2-55" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(<span class="fu">grepl</span>(<span class="st">"Campylobacter_D jejuni|Campylobacter_D coli"</span>, Species))</span>
<span id="cb2-56"><a href="#cb2-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-57"><a href="#cb2-57" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_delim</span>(<span class="at">x =</span> campylobacter_metadata,</span>
<span id="cb2-58"><a href="#cb2-58" aria-hidden="true" tabindex="-1"></a>              <span class="at">file =</span> campylobacter_metadata_file,</span>
<span id="cb2-59"><a href="#cb2-59" aria-hidden="true" tabindex="-1"></a>              <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb2-60"><a href="#cb2-60" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-61"><a href="#cb2-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-62"><a href="#cb2-62" aria-hidden="true" tabindex="-1"></a>campylobacter_species_numbers <span class="ot">&lt;-</span> campylobacter_metadata <span class="sc">%&gt;%</span></span>
<span id="cb2-63"><a href="#cb2-63" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">species =</span> <span class="fu">if_else</span>(<span class="fu">grepl</span>(<span class="st">"jejuni"</span>, Species), <span class="st">"jejuni"</span>, <span class="st">"coli"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb2-64"><a href="#cb2-64" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(species)</span>
<span id="cb2-65"><a href="#cb2-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-66"><a href="#cb2-66" aria-hidden="true" tabindex="-1"></a><span class="co"># Alternative visualisation method: stacked bar chart</span></span>
<span id="cb2-67"><a href="#cb2-67" aria-hidden="true" tabindex="-1"></a>stacked_species_plot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> campylobacter_species_numbers,</span>
<span id="cb2-68"><a href="#cb2-68" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">aes</span>(<span class="at">x =</span> <span class="dv">0</span>, <span class="at">y =</span> n, <span class="at">fill =</span> species)) <span class="sc">+</span></span>
<span id="cb2-69"><a href="#cb2-69" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb2-70"><a href="#cb2-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb2-71"><a href="#cb2-71" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="fu">paste</span>(<span class="st">"Total genomes:"</span>,</span>
<span id="cb2-72"><a href="#cb2-72" aria-hidden="true" tabindex="-1"></a>                <span class="fu">format</span>(<span class="fu">sum</span>(campylobacter_species_numbers<span class="sc">$</span>n),</span>
<span id="cb2-73"><a href="#cb2-73" aria-hidden="true" tabindex="-1"></a>                       <span class="at">big.mark =</span> <span class="st">","</span>))) <span class="sc">+</span></span>
<span id="cb2-74"><a href="#cb2-74" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-75"><a href="#cb2-75" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-76"><a href="#cb2-76" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb2-77"><a href="#cb2-77" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb2-78"><a href="#cb2-78" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Species"</span>,</span>
<span id="cb2-79"><a href="#cb2-79" aria-hidden="true" tabindex="-1"></a>                    <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#895d45"</span>, <span class="st">"#c6903c"</span>)) <span class="sc">+</span></span>
<span id="cb2-80"><a href="#cb2-80" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>() <span class="sc">+</span></span>
<span id="cb2-81"><a href="#cb2-81" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.01</span>))) <span class="sc">+</span></span>
<span id="cb2-82"><a href="#cb2-82" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">label =</span> n <span class="sc">%&gt;%</span></span>
<span id="cb2-83"><a href="#cb2-83" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">format</span>(<span class="at">big.mark =</span> <span class="st">","</span>)),</span>
<span id="cb2-84"><a href="#cb2-84" aria-hidden="true" tabindex="-1"></a>            <span class="at">position =</span> <span class="fu">position_stack</span>(<span class="at">vjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb2-85"><a href="#cb2-85" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="fl">4.5</span>,</span>
<span id="cb2-86"><a href="#cb2-86" aria-hidden="true" tabindex="-1"></a>            <span class="at">color =</span> <span class="st">"black"</span>)</span>
<span id="cb2-87"><a href="#cb2-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-88"><a href="#cb2-88" aria-hidden="true" tabindex="-1"></a>stacked_species_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-number_of_species" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-number_of_species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Descriptive_statistics_files/figure-html/fig-number_of_species-1.png" class="img-fluid figure-img" width="480">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-number_of_species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Number of species present in dataset
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="crispr-arrays-per-genome" class="level1">
<h1>CRISPR arrays per genome</h1>
<p>We have identified CRISPRs in all <em>Campylobacter jejuni</em> and <em>C. coli</em> genomes using <a href="https://github.com/Russel88/CRISPRCasTyper" title="GitHub repository for CRISPRCasTyper">CCTyper</a>. CCTyper separately annotates CRISPRs near <em>Cas</em> genes and separate CRISPR arrays, called ‘orphans’. To inventory the prevalence and size of both types of CRISPR arrays, show for each genome the number of CRISPR arrays as a scatterplot (<a href="#fig-crisprs_per_genome" class="quarto-xref">Figure&nbsp;2</a>). (Derived from the <code>CRISPR</code> field in CCTyper’s output.)</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>annotated_crispr_number_df <span class="ot">&lt;-</span> <span class="fu">left_join</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> campylobacter_metadata <span class="sc">%&gt;%</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(sample_accession, country, host, isolation_source,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>           scientific_name, tax_id, Species, description),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">full_join</span>(</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> crisprs_near_cas_df <span class="sc">%&gt;%</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">group_by</span>(sample_accession) <span class="sc">%&gt;%</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">count</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rename</span>(<span class="st">"CRISPRs near Cas"</span> <span class="ot">=</span> <span class="st">"n"</span>),</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> crisprs_orphan_df <span class="sc">%&gt;%</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">group_by</span>(sample_accession) <span class="sc">%&gt;%</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">count</span>() <span class="sc">%&gt;%</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">rename</span>(<span class="st">"Orphan CRISPRs"</span> <span class="ot">=</span> <span class="st">"n"</span>),</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"sample_accession"</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">across</span>(<span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span><span class="sc">:</span><span class="st">`</span><span class="at">Orphan CRISPRs</span><span class="st">`</span>,</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>           <span class="sc">~</span><span class="fu">replace_na</span>(., <span class="dv">0</span>)),</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_species =</span> <span class="fu">if_else</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"jejuni"</span>, <span class="at">x =</span> Species),</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"jejuni"</span>, <span class="st">"coli"</span>)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>number_of_crisprs_scatterplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> annotated_crispr_number_df,</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>,</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> <span class="st">`</span><span class="at">Orphan CRISPRs</span><span class="st">`</span>)</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(),</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="fl">0.7</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>()</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>crispr_near_cas_density <span class="ot">=</span> <span class="fu">ggplot</span>(</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> annotated_crispr_number_df,</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> <span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">fill =</span> <span class="st">"grey20"</span>) <span class="sc">+</span></span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>orphan_crispr_density <span class="ot">=</span> <span class="fu">ggplot</span>(</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> annotated_crispr_number_df,</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">y =</span> <span class="st">`</span><span class="at">Orphan CRISPRs</span><span class="st">`</span>)) <span class="sc">+</span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">fill =</span> <span class="st">"grey20"</span>) <span class="sc">+</span></span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>()</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="st">"</span></span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a><span class="st">1111#</span></span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a><span class="st">22223</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="st">22223</span></span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a><span class="st">"</span></span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a>crisprs_per_genome_figure <span class="ot">&lt;-</span> crispr_near_cas_density <span class="sc">+</span></span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>  number_of_crisprs_scatterplot <span class="sc">+</span></span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>  orphan_crispr_density <span class="sc">+</span></span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">design =</span> design,</span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>              <span class="at">axes =</span> <span class="st">"collect"</span>)</span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>crisprs_per_genome_figure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-crisprs_per_genome" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-crisprs_per_genome-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Descriptive_statistics_files/figure-html/fig-crisprs_per_genome-1.png" class="img-fluid figure-img" width="720">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crisprs_per_genome-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Number of CRISPR arrays per genome
</figcaption>
</figure>
</div>
</div>
</div>
<p>From this figure, it appears that by far the most genomes have no CRISPR arrays at all, and otherwise they mostly have one CRISPR array near <em>Cas</em> genes (one contig containing CRISPRs). At most, CCTyper has found 3 CRISPRs near <em>Cas</em> genes and up to 4 orphan CRISPRs Let’s see this distribution separately between <em>C. jejuni</em> and <em>coli</em>, and look at the number of spacers per genome.</p>
<section id="arrays-per-genome-and-per-species" class="level2">
<h2 class="anchored" data-anchor-id="arrays-per-genome-and-per-species">Arrays per genome and per species</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>total_genomes <span class="ot">=</span> <span class="fu">nrow</span>(annotated_crispr_number_df)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>annotated_crispr_number_df <span class="sc">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(<span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>, <span class="at">name =</span> <span class="st">"Genomes"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Percentage =</span> <span class="fu">round</span>(Genomes <span class="sc">/</span> total_genomes <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["CRISPRs near Cas"],"name":[1],"type":["int"],"align":["right"]},{"label":["Genomes"],"name":[2],"type":["int"],"align":["right"]},{"label":["Percentage"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"0","2":"8443","3":"43.42"},{"1":"1","2":"10744","3":"55.25"},{"1":"2","2":"256","3":"1.32"},{"1":"3","2":"2","3":"0.01"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>prepare_stats_df <span class="ot">&lt;-</span> <span class="cf">function</span>(numbers_df, stat, <span class="at">species =</span> <span class="st">"*"</span>) {</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(numbers_df <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>           <span class="fu">filter</span>(<span class="fu">grepl</span>(<span class="at">pattern =</span> species, <span class="at">x =</span> Species)) <span class="sc">%&gt;%</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>           <span class="fu">group_by</span>({{ stat }}) <span class="sc">%&gt;%</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>           <span class="fu">summarise</span>(<span class="at">Genomes =</span> <span class="fu">length</span>({{ stat }})) <span class="sc">%&gt;%</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>           <span class="fu">mutate</span>(<span class="at">total_genomes =</span> <span class="fu">sum</span>(Genomes),</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">Percentage =</span> <span class="fu">round</span>(Genomes <span class="sc">/</span> total_genomes <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">2</span>),</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ymax =</span> <span class="fu">cumsum</span>(Percentage),</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ymin =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">head</span>(ymax, <span class="at">n=</span><span class="sc">-</span><span class="dv">1</span>)),</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">label_position =</span> (ymax <span class="sc">+</span> ymin) <span class="sc">/</span> <span class="dv">2</span>,</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>                  <span class="at">label =</span> <span class="fu">paste0</span>({{ stat }}, <span class="st">" CRISPRs:</span><span class="sc">\n</span><span class="st">"</span>, Percentage, <span class="st">"%"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>           <span class="fu">select</span>(<span class="sc">-</span>total_genomes)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>make_donut_chart <span class="ot">&lt;-</span> <span class="cf">function</span>(stats_df) {</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>      <span class="at">data =</span> stats_df,</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">ymax =</span> ymax, <span class="at">ymin =</span> ymin,</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>                    <span class="at">xmax =</span> <span class="dv">4</span>, <span class="at">xmin =</span> <span class="dv">3</span>,</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fill =</span> <span class="sc">-</span>label_position)) <span class="sc">+</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_rect</span>() <span class="sc">+</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>      <span class="fu">coord_polar</span>(<span class="at">theta =</span> <span class="st">"y"</span>) <span class="sc">+</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">4</span>)) <span class="sc">+</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_text</span>(<span class="at">x =</span> <span class="fl">4.2</span>, <span class="fu">aes</span>(<span class="at">label =</span> label, <span class="at">y =</span> label_position),</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>                <span class="at">size =</span> <span class="dv">3</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.3</span>)) <span class="sc">+</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"CRISPRs in Campylobacter jejuni"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CRISPRs in Campylobacter jejuni"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>crisprs_jejuni_df <span class="ot">&lt;-</span> <span class="fu">prepare_stats_df</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">numbers_df =</span> annotated_crispr_number_df,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">stat =</span> <span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="st">"Campylobacter_D jejuni"</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>crisprs_jejuni_df[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["CRISPRs near Cas"],"name":[1],"type":["int"],"align":["right"]},{"label":["Genomes"],"name":[2],"type":["int"],"align":["right"]},{"label":["Percentage"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"0","2":"4493","3":"32.39"},{"1":"1","2":"9149","3":"65.95"},{"1":"2","2":"231","3":"1.67"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The number of other jejuni subspecies(?) is too low to do</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># separate analyses, so just put them all together</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Orphan CRISPRs in C. jejuni"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Orphan CRISPRs in C. jejuni"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>orphan_crisprs_jejuni <span class="ot">&lt;-</span> <span class="fu">prepare_stats_df</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">numbers_df =</span> annotated_crispr_number_df,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">stat =</span> <span class="st">`</span><span class="at">Orphan CRISPRs</span><span class="st">`</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="st">"Campylobacter_D jejuni"</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>orphan_crisprs_jejuni[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Orphan CRISPRs"],"name":[1],"type":["int"],"align":["right"]},{"label":["Genomes"],"name":[2],"type":["int"],"align":["right"]},{"label":["Percentage"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"0","2":"13319","3":"96.01"},{"1":"1","2":"539","3":"3.89"},{"1":"2","2":"15","3":"0.11"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"CRISPRs in Campylobacter coli"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CRISPRs in Campylobacter coli"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>crisprs_coli <span class="ot">&lt;-</span> <span class="fu">prepare_stats_df</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">numbers_df =</span> annotated_crispr_number_df,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">stat =</span> <span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="st">"Campylobacter_D coli"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>crisprs_coli[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["CRISPRs near Cas"],"name":[1],"type":["int"],"align":["right"]},{"label":["Genomes"],"name":[2],"type":["int"],"align":["right"]},{"label":["Percentage"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"0","2":"3950","3":"70.89"},{"1":"1","2":"1595","3":"28.63"},{"1":"2","2":"25","3":"0.45"},{"1":"3","2":"2","3":"0.04"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Orphan CRISPRs in C. coli"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Orphan CRISPRs in C. coli"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>orphan_crisprs_coli <span class="ot">&lt;-</span> <span class="fu">prepare_stats_df</span>(</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">numbers_df =</span> annotated_crispr_number_df,</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">stat =</span> <span class="st">`</span><span class="at">Orphan CRISPRs</span><span class="st">`</span>,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">species =</span> <span class="st">"Campylobacter_D coli"</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>orphan_crisprs_coli[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Orphan CRISPRs"],"name":[1],"type":["int"],"align":["right"]},{"label":["Genomes"],"name":[2],"type":["int"],"align":["right"]},{"label":["Percentage"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"0","2":"5390","3":"96.73"},{"1":"1","2":"170","3":"3.05"},{"1":"2","2":"10","3":"0.18"},{"1":"3","2":"1","3":"0.02"},{"1":"4","2":"1","3":"0.02"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>crisprs_near_cas_figure <span class="ot">&lt;-</span> <span class="fu">make_donut_chart</span>(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">prepare_stats_df</span>(<span class="at">numbers_df =</span> annotated_crispr_number_df,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">stat =</span> <span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Overall"</span>) <span class="sc">+</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">make_donut_chart</span>(crisprs_jejuni_df) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"jejuni"</span>) <span class="sc">+</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">make_donut_chart</span>(crisprs_coli) <span class="sc">+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"coli"</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>crisprs_near_cas_figure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-crisprs_per_genome_donut" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-crisprs_per_genome_donut-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Descriptive_statistics_files/figure-html/fig-crisprs_per_genome_donut-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-crisprs_per_genome_donut-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Prevalence of CRISPR arrays
</figcaption>
</figure>
</div>
</div>
</div>
<p>That is a visual and numerical impression of the prevalence of CRISPRs in <em>C. jejuni</em> and <em>C. coli</em> genomes. It seems as though the prevalence is quite different between species. Let’s test that statistically.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>crisprs_per_species_table <span class="ot">&lt;-</span> annotated_crispr_number_df <span class="sc">%&gt;%</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(<span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>, simple_species) <span class="sc">%&gt;%</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">genomes =</span> <span class="fu">length</span>(<span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xtabs</span>(<span class="at">formula =</span> genomes <span class="sc">~</span> <span class="st">`</span><span class="at">CRISPRs near Cas</span><span class="st">`</span> <span class="sc">+</span> simple_species, <span class="at">data =</span> .)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>crisprs_per_species_table</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                simple_species
CRISPRs near Cas coli jejuni
               0 3950   4493
               1 1595   9149
               2   25    231
               3    2      0</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>crisprs_per_species_test <span class="ot">&lt;-</span> <span class="fu">prop_test</span>(</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> crisprs_per_species_table, <span class="at">detailed =</span> T)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>crisprs_per_species_test</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["n"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["n1"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["n2"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["n3"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["n4"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["estimate1"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["estimate2"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["estimate3"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["estimate4"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["statistic"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["p"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["df"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["method"],"name":[13],"type":["chr"],"align":["left"]},{"label":["alternative"],"name":[14],"type":["chr"],"align":["left"]},{"label":["p.signif"],"name":[15],"type":["chr"],"align":["left"]}],"data":[{"1":"19445","2":"8443","3":"10744","4":"256","5":"2","6":"0.4678432","7":"0.148455","8":"0.09765625","9":"1","10":"2409.221","11":"0","12":"3","13":"Prop test","14":"two.sided","15":"****","_rn_":"1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Two-sided <span class="math inline">\(\chi^2\)</span> shows a significant difference in prevalence of CRISPR arrays between <em>C. jejuni</em> and <em>C. coli</em> (<span class="math inline">\(\chi\)</span> = 2409, df = 3, p = 0). This may be loosely translated into: <em>C. jejuni</em> genomes more frequently have CRISPRs than <em>C. coli</em>.</p>
<p>Now let’s see the predicted type of the CRISPR arrays (using the column <code>Prediction</code> from the (concatenated) files <code>crisprs_near_cas.tab</code> and <code>crisprs_orphan.tab</code>).</p>
</section>
<section id="type-of-crispr-arrays" class="level2">
<h2 class="anchored" data-anchor-id="type-of-crispr-arrays">Type of CRISPR arrays</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>selected_metadata <span class="ot">&lt;-</span> campylobacter_metadata <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(sample_accession, country, host, isolation_source,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>           scientific_name, tax_id, Species, description)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>annotated_crispr_df <span class="ot">&lt;-</span> <span class="fu">rbind</span>(</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> selected_metadata,</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> crisprs_near_cas_df,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"sample_accession"</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">orientation =</span> <span class="st">"near_Cas"</span>),</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> selected_metadata,</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> crisprs_orphan_df,</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"sample_accession"</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">orientation =</span> <span class="st">"orphan"</span>)</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">simple_species =</span> <span class="fu">if_else</span>(</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">grepl</span>(<span class="at">pattern =</span> <span class="st">"jejuni"</span>, <span class="at">x =</span> Species),</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>    <span class="st">"jejuni"</span>, <span class="st">"coli"</span>),</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">N_spacers =</span> <span class="fu">case_when</span>(</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>      N_repeats <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">~</span> N_repeats <span class="sc">-</span> <span class="dv">1</span>,</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>      <span class="fu">is.na</span>(N_repeats) <span class="sc">~</span> <span class="dv">0</span>,</span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> <span class="dv">0</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>annotated_crispr_df <span class="sc">%&gt;%</span></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(orientation, Prediction) <span class="sc">%&gt;%</span></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(orientation, <span class="fu">desc</span>(n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["orientation"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Prediction"],"name":[2],"type":["chr"],"align":["left"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"near_Cas","2":"II-C","3":"10846"},{"1":"near_Cas","2":"NA","3":"8443"},{"1":"near_Cas","2":"Unknown","3":"392"},{"1":"near_Cas","2":"I-B","3":"18"},{"1":"near_Cas","2":"II-A","3":"4"},{"1":"near_Cas","2":"V-F1","3":"2"},{"1":"orphan","2":"NA","3":"18709"},{"1":"orphan","2":"II-C","3":"538"},{"1":"orphan","2":"Unknown","3":"218"},{"1":"orphan","2":"I-B","3":"7"},{"1":"orphan","2":"II-A","3":"3"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>annotated_crispr_df <span class="sc">%&gt;%</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(simple_species, orientation, Prediction) <span class="sc">%&gt;%</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(simple_species), orientation, <span class="fu">desc</span>(n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["simple_species"],"name":[1],"type":["chr"],"align":["left"]},{"label":["orientation"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Prediction"],"name":[3],"type":["chr"],"align":["left"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"jejuni","2":"near_Cas","3":"II-C","4":"9268"},{"1":"jejuni","2":"near_Cas","3":"NA","4":"4493"},{"1":"jejuni","2":"near_Cas","3":"Unknown","4":"327"},{"1":"jejuni","2":"near_Cas","3":"I-B","4":"15"},{"1":"jejuni","2":"near_Cas","3":"II-A","4":"1"},{"1":"jejuni","2":"orphan","3":"NA","4":"13319"},{"1":"jejuni","2":"orphan","3":"II-C","4":"396"},{"1":"jejuni","2":"orphan","3":"Unknown","4":"169"},{"1":"jejuni","2":"orphan","3":"I-B","4":"4"},{"1":"coli","2":"near_Cas","3":"NA","4":"3950"},{"1":"coli","2":"near_Cas","3":"II-C","4":"1578"},{"1":"coli","2":"near_Cas","3":"Unknown","4":"65"},{"1":"coli","2":"near_Cas","3":"I-B","4":"3"},{"1":"coli","2":"near_Cas","3":"II-A","4":"3"},{"1":"coli","2":"near_Cas","3":"V-F1","4":"2"},{"1":"coli","2":"orphan","3":"NA","4":"5390"},{"1":"coli","2":"orphan","3":"II-C","4":"142"},{"1":"coli","2":"orphan","3":"Unknown","4":"49"},{"1":"coli","2":"orphan","3":"I-B","4":"3"},{"1":"coli","2":"orphan","3":"II-A","4":"3"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>Among the both the CRISPR arrays with <em>Cas</em> genes nearby and the orphan CRISPRs the most prevalent type predicted is II-C. The second most prevalent type is ‘unknown’, which might be II-C with low certainty. This pattern occurs in both <em>C. jejuni</em> and <em>C. coli</em>. (Note: ‘Prediction = NA’ refers to absence of CRISPRs, genomes with no CRISPRs ‘near_Cas’ or ‘orphan’.)</p>
<p>Now that we have seen the prevalence of CRISPR arrays and their type, let’s have a closer look at the number of CRISPR spacers and repeats within each array.</p>
</section>
</section>
<section id="crispr-spacers-per-genome" class="level1">
<h1>CRISPR spacers per genome</h1>
<p>From a little manual inspection, I found that CCTyper reports one more repeat per genome than spacers (as counted from the entries in the fasta files generated). Therefore, number of spacers are reported as number of repeats (<code>N_repeats</code>) minus one.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>total_spacer_scatter_data <span class="ot">&lt;-</span> annotated_crispr_df <span class="sc">%&gt;%</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> <span class="fu">c</span>(sample_accession, simple_species),</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">names_from =</span> orientation,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_from =</span> N_spacers,</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">values_fn =</span> sum)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>number_of_spacers_scatterplot <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> total_spacer_scatter_data,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> near_Cas,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> orphan)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(),</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>             <span class="at">size =</span> <span class="dv">1</span>, <span class="at">alpha =</span> <span class="fl">0.3</span>) <span class="sc">+</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"CRISPR spacers near Cas"</span>,</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Orphan CRISPR spacers"</span>)</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>spacers_near_cas_density <span class="ot">=</span> <span class="fu">ggplot</span>(</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> total_spacer_scatter_data,</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> near_Cas)) <span class="sc">+</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">fill =</span> <span class="st">"grey20"</span>) <span class="sc">+</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>orphan_spacer_density <span class="ot">=</span> <span class="fu">ggplot</span>(</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> total_spacer_scatter_data,</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">y =</span> orphan)) <span class="sc">+</span></span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">fill =</span> <span class="st">"grey20"</span>) <span class="sc">+</span></span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_blank</span>())</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a>spacers_per_genome_figure <span class="ot">&lt;-</span> spacers_near_cas_density <span class="sc">+</span></span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a>  number_of_spacers_scatterplot <span class="sc">+</span></span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a>  orphan_spacer_density <span class="sc">+</span></span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">design =</span> design,</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>              <span class="at">axes =</span> <span class="st">"collect"</span>)</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a>spacers_per_genome_figure</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-spacers_per_genome" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-spacers_per_genome-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Descriptive_statistics_files/figure-html/fig-spacers_per_genome-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-spacers_per_genome-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Total number of CRISPR spacers
</figcaption>
</figure>
</div>
</div>
</div>
<p>This figure shows (again) that orphan CRISPRs are relatively rare, but when they occur they may contain up to 23 spacers. The most commonly occurring number of spacers is 3. As for the CRISPRs with <em>Cas</em> genes nearby, most CRISPR arrays contain up to 13 spacers (each at least 100 times) with higher numbers being observed less often. The minimum number of spacers is 2, and the max 45. (Note that these numbers sum all spacers if there are multiple arrays!)</p>
<section id="spacers-per-species" class="level2">
<h2 class="anchored" data-anchor-id="spacers-per-species">Spacers per species</h2>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>annotated_crispr_df <span class="sc">%&gt;%</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> N_repeats <span class="sc">-</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="fu">vars</span>(simple_species, orientation),</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">scales =</span> <span class="st">"free_y"</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">labeller =</span> <span class="fu">labeller</span>(</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>               <span class="at">simple_species =</span> <span class="fu">c</span>(<span class="at">coli =</span> <span class="st">"C. coli"</span>, <span class="at">jejuni =</span> <span class="st">"C. jejuni"</span>),</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>               <span class="at">orientation =</span> <span class="fu">c</span>(<span class="at">near_Cas =</span> <span class="st">"Near Cas"</span>, <span class="at">orphan =</span> <span class="st">"Orphan"</span>)),</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>             <span class="at">axes =</span> <span class="st">"all_x"</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>             )<span class="sc">+</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>))) <span class="sc">+</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of spacers"</span>,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-total_spacers_per_species" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-total_spacers_per_species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Descriptive_statistics_files/figure-html/fig-total_spacers_per_species-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-total_spacers_per_species-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Total spacers per species
</figcaption>
</figure>
</div>
</div>
</div>
<p>There appears to be a difference in distribution of the number of CRISPR spacers between <em>C. jejuni</em> (bottom) and <em>C. coli</em> (top). In <em>C. jejuni</em> there is a clear anticorrelation between number of spacers and frequency: most arrays harbour 2 spacers and for each added spacer the frequency decreases. This pattern does not apply to <em>C. coli</em>: the most common <em>Cas</em>-containing array has 5 spacers and frequencies of arrays with 11 and 13 spacers are higher than those for 10, 12, 14 or more spacers. This might be correlated with other underlying biological data, such as the origin of the genomes (different environments, or lineages?).</p>
</section>
</section>
<section id="length-of-spacers-and-repeats" class="level1">
<h1>Length of spacers and repeats</h1>
<p>CCTyper reports for each array the length of the repeats and the mean length of the CRISPR spacers with standard error of the mean (SEM; as <code>Repeat_len</code>, <code>Spacer_len_avg</code>, and <code>Spacer_len_sem</code>). From previous tests and discussions it appears that these may shift slightly between genomes, while the total length remains the same, meaning that for instance a repeat may be 3bp longer and the spacer 3bp shorter compared to a similar array in another genome. Let’s make a correlation dot plot to show this.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>repeat_and_spacer_length_figure <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> annotated_crispr_df,</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Spacer_len_avg,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> Repeat_len)</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.8</span>, <span class="at">alpha =</span> <span class="fl">0.8</span>) <span class="sc">+</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_cor</span>(<span class="at">method =</span> <span class="st">"pearson"</span>, <span class="at">label.x.npc =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Spacer length (bp)"</span>,</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Repeat length (bp)"</span>)</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>repeat_length_histogram <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>  annotated_crispr_df,</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">y =</span> Repeat_len)) <span class="sc">+</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="cn">NULL</span>,</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Count"</span>)</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>spacer_length_histogram <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>  annotated_crispr_df,</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Spacer_len_avg)) <span class="sc">+</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_pubr</span>() <span class="sc">+</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="cn">NULL</span>,</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Count"</span>)</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>spacer_length_histogram <span class="sc">+</span></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_spacer</span>() <span class="sc">+</span></span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>  repeat_and_spacer_length_figure <span class="sc">+</span></span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>  repeat_length_histogram <span class="sc">+</span></span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">1</span>),</span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>              <span class="at">heights =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-repeat_and_spacer_length" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-repeat_and_spacer_length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="Descriptive_statistics_files/figure-html/fig-repeat_and_spacer_length-1.png" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-repeat_and_spacer_length-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: CRISPR repeat and spacer length
</figcaption>
</figure>
</div>
</div>
</div>
<p>Indeed, lengths of repeats and spacers are highly correlated (Pearson’s R = -0.99, p &lt; 0.0001).The most commonly found combination is spacers of 30 bp long with repeats of 36 bp. It is likely that those are then the canonical lengths and deviations - especially those where the combined length is still 66 - may be the results of technical inaccuracies. Some combinations have different combined lengths and are more likely to be biologically different.</p>
</section>
<section id="repeat-sequence-summary" class="level1">
<h1>Repeat sequence summary</h1>
<p>Look at the consensus repeat sequences to get an idea of the diversity and possible errors in calling.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>annotated_crispr_df <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(Consensus_repeat, Repeat_len) <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Consensus_repeat"],"name":[1],"type":["chr"],"align":["left"]},{"label":["Repeat_len"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["n"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"NA","2":"NA","3":"27152"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.0","3":"3129"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"36.0","3":"2940"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAA","2":"35.0","3":"860"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"35.0","3":"812"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"37.0","3":"401"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"37.0","3":"286"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.0","3":"221"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"36.0","3":"201"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"37.0","3":"178"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.0","3":"173"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"37.0","3":"114"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.0","3":"114"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"36.0","3":"113"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATGATAAAATA","2":"37.0","3":"108"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAA","2":"36.0","3":"107"},{"1":"TATTTTATCATATAAACAATTAATAATAGGCTAAAAC","2":"37.0","3":"87"},{"1":"CCCTTTTTAAATTTCTTTATGGTAAAAT","2":"28.0","3":"80"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAA","2":"33.0","3":"76"},{"1":"TTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"33.0","3":"64"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGG","2":"28.0","3":"63"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"38.0","3":"61"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"37.0","3":"60"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"37.0","3":"58"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"38.0","3":"55"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"37.0","3":"50"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGAGACTAAAAC","2":"36.0","3":"49"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"38.0","3":"46"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"37.0","3":"45"},{"1":"GTTTTAGTCTCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.0","3":"36"},{"1":"TTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"38.0","3":"30"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"38.0","3":"29"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAA","2":"38.0","3":"28"},{"1":"ATTTTACCATATAAACAATTAATAATAGGCTAAAAC","2":"36.0","3":"27"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATGATAAAAT","2":"36.0","3":"27"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"36.8","3":"26"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATGGTAAAAT","2":"36.0","3":"26"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"38.0","3":"25"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATG","2":"29.0","3":"24"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTAA","2":"39.0","3":"23"},{"1":"ATTTTATCATATAAACAATTAATAATAGGCTAAAAC","2":"36.0","3":"22"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"38.0","3":"22"},{"1":"TTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"34.0","3":"22"},{"1":"GTTTTAGCCTATTATTAATTGTTT","2":"24.0","3":"21"},{"1":"TTTTAGCCTATTATTAATTGTTTATATGATAAAAT","2":"35.0","3":"21"},{"1":"CATATAAACAATTAATAATAGGCTAAAAC","2":"29.0","3":"20"},{"1":"GTTTTAGCCTATTATTAATTATTTATATGATAAAAT","2":"36.0","3":"19"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAA","2":"34.0","3":"18"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACC","2":"37.0","3":"16"},{"1":"TTAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"15"},{"1":"GGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.0","3":"14"},{"1":"GTTTTAGCCTATTGTTAATTGTTTGTATGGTAAAAT","2":"36.0","3":"14"},{"1":"TATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.0","3":"14"},{"1":"ATTTTATCATACAAACAATTAATAATAGGCTAAAAC","2":"36.0","3":"13"},{"1":"GTTTTAGCCTATTATTAATTGTTTGTATGATAAAAT","2":"36.0","3":"13"},{"1":"TGTTTTAGCCTATTATTAATTATTTATATGATAAAAT","2":"37.0","3":"13"},{"1":"TGTTTTAGCCTATTATTAATTGTTTATATG","2":"30.0","3":"13"},{"1":"ATTTTATCATATAAACAATTAATAATAGGCTAAAACT","2":"37.0","3":"12"},{"1":"ATTTTATCATATAAATAATTAATAATAGGCTAAAACA","2":"37.0","3":"12"},{"1":"GTTTTAACCTATTATTAATTGTTTATATGATAAAAT","2":"36.0","3":"12"},{"1":"GTTTTAGCCTATTGTTAATTGTTTGTATGATAAAAT","2":"36.0","3":"12"},{"1":"AGTTTTAGCCTATTATTAATTGTTTATATGATAAAAT","2":"37.0","3":"11"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTA","2":"38.0","3":"11"},{"1":"ATTTTATCATACAAACAATTAACAATAGGCTAAAAC","2":"36.0","3":"11"},{"1":"CATATAAACAATTAATAATAGGCTAAAACA","2":"30.0","3":"11"},{"1":"GTTTTAGCCTATTATTAATTGTTTACATGGTAAAAT","2":"36.0","3":"11"},{"1":"TTTTAGTCCCTTTTTAAATTTCTT","2":"24.0","3":"11"},{"1":"AAACAATTAATAATAGGCTAAAAC","2":"24.0","3":"10"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATA","2":"37.0","3":"10"},{"1":"ATTTTACCATATAAACAATTAATAATAGGCTAAAACA","2":"37.0","3":"10"},{"1":"TGTTTTAGCCTATTATTAATTGTTTATATGGTAAAAT","2":"37.0","3":"10"},{"1":"AAGAAATTTAAAAAGGGACTAAAA","2":"24.0","3":"9"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.7","3":"9"},{"1":"ATTTTACCATACAAACAATTAACAATAGGCTAAAAC","2":"36.0","3":"9"},{"1":"ATTTTACCATATAAACAATTAATAATAGGCTAAAAC","2":"35.7","3":"9"},{"1":"ATTTTATCATATAAACAATTAATAATAGGTTAAAACA","2":"37.0","3":"9"},{"1":"ATTTTATCATATAAATAATTAATAATAGGCTAAAAC","2":"36.0","3":"9"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"36.7","3":"9"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAG","2":"38.0","3":"9"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"36.7","3":"9"},{"1":"TTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"33.0","3":"9"},{"1":"TTATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"9"},{"1":"ATTTTATCATATAAACAATTAATAATAGGCTAAAA","2":"35.0","3":"8"},{"1":"TGTTTTAACCTATTATTAATTGTTTATATGATAAAAT","2":"37.0","3":"8"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATAA","2":"38.0","3":"7"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAG","2":"37.0","3":"7"},{"1":"ATTTTATCATATAAACAATTAATAATAGGTTAAAAC","2":"36.0","3":"6"},{"1":"CATTTTACCATAAAGAAATTTAAAA","2":"25.0","3":"6"},{"1":"CTATTATTAATTGTTTATATGATAAAATA","2":"29.0","3":"6"},{"1":"GTTTTAGCCTATTATTAATTGTTTA","2":"25.0","3":"6"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAT","2":"38.0","3":"6"},{"1":"TAAACAATTAATAATAGGCTAAAAC","2":"25.0","3":"6"},{"1":"TAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"38.0","3":"6"},{"1":"TTATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"38.0","3":"6"},{"1":"TTTTTAAATTTCTTTATGGTAAAAT","2":"25.0","3":"6"},{"1":"AGTTTTAGCCTATTATTAATTGTTTATATGATAAAATA","2":"38.0","3":"5"},{"1":"ATATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"38.0","3":"5"},{"1":"ATTTTACCATAAAGAAATTGAAAAAGGGACTAAAAC","2":"36.7","3":"5"},{"1":"ATTTTACCATGTAAACAATTAATAATAGGCTAAAAC","2":"36.0","3":"5"},{"1":"ATTTTATCATATAAACAATTAACAATAGGCTAAAA","2":"35.7","3":"5"},{"1":"CAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"39.0","3":"5"},{"1":"CTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"38.0","3":"5"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATGGTAAAAT","2":"35.7","3":"5"},{"1":"GTTTTAGCCTATTGTTAATTGTTTATATGGTAAAAT","2":"35.7","3":"5"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAA","2":"33.0","3":"5"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAA","2":"39.0","3":"5"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATT","2":"38.0","3":"5"},{"1":"TGTTTTAGTCTCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.0","3":"5"},{"1":"TTTTAAATTTCTTTATGGTAAAAT","2":"24.0","3":"5"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAA","2":"34.0","3":"4"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.0","3":"4"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAA","2":"39.0","3":"4"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAG","2":"39.0","3":"4"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAAC","2":"39.0","3":"4"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAAT","2":"39.0","3":"4"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTAC","2":"39.0","3":"4"},{"1":"ATTTTACCATAAAGGAATTTAAAAAGGGACTAAAACT","2":"37.0","3":"4"},{"1":"CATTTTACCATAAAGAAATTTAAAAAGGGACTAAAA","2":"36.0","3":"4"},{"1":"CTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAA","2":"37.0","3":"4"},{"1":"CTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"39.0","3":"4"},{"1":"TAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"39.0","3":"4"},{"1":"TATTTTACCATAAAGAAAATTAAAAAGGGACTAA","2":"36.0","3":"4"},{"1":"TATTTTACCATAAAGAAATTTAAAA","2":"25.0","3":"4"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"36.0","3":"4"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATAA","2":"39.0","3":"4"},{"1":"TATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTG","2":"39.0","3":"4"},{"1":"TATTTTATCATATAAACAATTAATAATAG","2":"29.0","3":"4"},{"1":"TATTTTATCATATAAACAATTAATAATAGGCTAAAACT","2":"38.0","3":"4"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAA","2":"39.0","3":"4"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAG","2":"39.0","3":"4"},{"1":"TTAGTCCCTTTTTAATTTTCTTTATGGTAAAATA","2":"36.0","3":"4"},{"1":"TTTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"39.0","3":"4"},{"1":"TTTTAGCCTATTGTTAATTGTTTATATGATAAAAT","2":"35.7","3":"4"},{"1":"AAATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATCG","2":"40.0","3":"3"},{"1":"AAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"38.0","3":"3"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGAGACTAAAAC","2":"37.0","3":"3"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"36.8","3":"3"},{"1":"AATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.0","3":"3"},{"1":"ATCTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"36.0","3":"3"},{"1":"ATTTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"40.0","3":"3"},{"1":"ATTTTACCATAAAGAAATTTAAAA","2":"24.0","3":"3"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGAGACTAAAACA","2":"37.0","3":"3"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"36.7","3":"3"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTT","2":"38.0","3":"3"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATT","2":"37.0","3":"3"},{"1":"ATTTTACCATATAAACAATTAATAATAGGTTAAAAC","2":"36.0","3":"3"},{"1":"ATTTTACCATGTAAACAATTAATAATAGGCTAAAAC","2":"35.7","3":"3"},{"1":"CTATTTTACCATAAAGAAATTGAAAAAGGGACTAAAATAA","2":"40.0","3":"3"},{"1":"GTTTTAGCCTATTGTTAATTGTTTGTATGATAAAATA","2":"37.0","3":"3"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAAA","2":"39.0","3":"3"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAAAT","2":"40.0","3":"3"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAATG","2":"40.0","3":"3"},{"1":"GTTTTAGTCCCTTTTTCAATTTCTTTATGGTAAAAT","2":"36.7","3":"3"},{"1":"TATTATTAATTGTTTACATGGTAAAAT","2":"27.0","3":"3"},{"1":"TATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"38.0","3":"3"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.2","3":"3"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATC","2":"38.0","3":"3"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"37.2","3":"3"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTA","2":"39.0","3":"3"},{"1":"TTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"39.0","3":"3"},{"1":"TTATTTTAGTCCCTTTTTCAATTTCTTTATGGTAAAATAG","2":"40.0","3":"3"},{"1":"TTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"3"},{"1":"TTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTTTGT","2":"40.0","3":"3"},{"1":"TTTTAGCCTATTATTAATTGTTTATATG","2":"28.0","3":"3"},{"1":"TTTTAGCCTATTGTTAATTGTTT","2":"23.0","3":"3"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATG","2":"36.0","3":"3"},{"1":"TTTTTAAATTTCTTTATGGTAAAATA","2":"26.0","3":"3"},{"1":"TTTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTA","2":"39.0","3":"3"},{"1":"AAACAATTAATAATAGGCTAAAACA","2":"25.0","3":"2"},{"1":"AAACAATTAATAATAGGCTAAAACT","2":"25.0","3":"2"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"37.2","3":"2"},{"1":"AGTTTTAGTCCCTTTTTAAATTCCTTTATGGTAAAAT","2":"37.0","3":"2"},{"1":"ATCTTACCATAAAGAAATTTAAAAAGGGA","2":"29.0","3":"2"},{"1":"ATGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"38.0","3":"2"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGAGA","2":"29.0","3":"2"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACCA","2":"38.0","3":"2"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"37.2","3":"2"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTTA","2":"39.0","3":"2"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.3","3":"2"},{"1":"ATTTTATCATACAAACAATTAACAATAGGCTAAAAC","2":"35.7","3":"2"},{"1":"ATTTTATCATACAAACATTTAATAAGAGGCTAAAAC","2":"36.0","3":"2"},{"1":"ATTTTATCATATAAACAATTAATAATAGG","2":"29.0","3":"2"},{"1":"ATTTTATCATATAAACAATTAATGATAGGCTAAAGC","2":"36.0","3":"2"},{"1":"ATTTTTAGCCTCTTGTTAATTGTTTGTATGATAAATT","2":"37.0","3":"2"},{"1":"CAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"39.0","3":"2"},{"1":"CAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATA","2":"39.0","3":"2"},{"1":"CATAAAGAAATTTAAAAAGGGACTAAAAC","2":"29.0","3":"2"},{"1":"CATACAAACAATTAACAATAGGCTAAAA","2":"28.0","3":"2"},{"1":"CATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"38.0","3":"2"},{"1":"CCCTTTTTAAATTTCTTTATGGTAAAATA","2":"29.0","3":"2"},{"1":"CTATTTTACCATATAAACAATTAATAATAGGCTAAAAC","2":"38.0","3":"2"},{"1":"CTTTTTAAATTTCTTTATGGTAAAAT","2":"26.0","3":"2"},{"1":"GAAAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"40.0","3":"2"},{"1":"GGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"38.0","3":"2"},{"1":"GTTTTAGCCTATTATTAATTGTTTACATGGTAAAAT","2":"35.7","3":"2"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATGGTAAAATAG","2":"38.0","3":"2"},{"1":"GTTTTAGCCTATTGTTAATTGTTT","2":"24.0","3":"2"},{"1":"GTTTTAGCCTATTGTTAATTGTTTATATGGTAAAAT","2":"36.0","3":"2"},{"1":"GTTTTAGCCTATTGTTAATTGTTTGTATGATAAAAT","2":"35.7","3":"2"},{"1":"GTTTTAGCCTCTTGTTAAATGTTTGTATGATAAAAT","2":"36.0","3":"2"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGATAAAAT","2":"36.0","3":"2"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATG","2":"37.0","3":"2"},{"1":"GTTTTAGTCTCTTTTTAAATTTCTTTATGGTAAAATT","2":"37.0","3":"2"},{"1":"TAAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"2"},{"1":"TATGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"40.0","3":"2"},{"1":"TATTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"40.0","3":"2"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGAGACTAAAAC","2":"37.0","3":"2"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATTT","2":"39.0","3":"2"},{"1":"TATTTTATCATACAAACAATTAACAATAGGCTAAAA","2":"36.0","3":"2"},{"1":"TCTAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"40.0","3":"2"},{"1":"TCTCTTTTTAAATTTCTTTATGGTAAAAT","2":"29.0","3":"2"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.7","3":"2"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.8","3":"2"},{"1":"TTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"34.0","3":"2"},{"1":"TTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"39.0","3":"2"},{"1":"TTTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"2"},{"1":"TTTTAGCCTATTGTTAATTGTTTGTATG","2":"28.0","3":"2"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAA","2":"34.0","3":"2"},{"1":"AAACAATTAACAAGAGACTAAAA","2":"23.0","3":"1"},{"1":"AAACAATTAACAATAGGCTAAAA","2":"23.0","3":"1"},{"1":"AAACAATTAACAATAGGCTAAAAC","2":"24.0","3":"1"},{"1":"AAAGAAATTTAAAAAGGGACTAAAAC","2":"26.0","3":"1"},{"1":"AAATATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAA","2":"40.0","3":"1"},{"1":"AAATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"1"},{"1":"AAATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAC","2":"40.0","3":"1"},{"1":"AAGTTTTAGCCTATTATTAATTGTTTATATGATAAAAT","2":"38.0","3":"1"},{"1":"AAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"1"},{"1":"AATATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"40.0","3":"1"},{"1":"AATGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"1"},{"1":"AATTTTACCATAAAGAAATTGAAAAAGGGACTAAAAC","2":"37.0","3":"1"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"37.0","3":"1"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"37.2","3":"1"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"37.8","3":"1"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAA","2":"39.0","3":"1"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAG","2":"39.0","3":"1"},{"1":"AATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"37.0","3":"1"},{"1":"AATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"38.0","3":"1"},{"1":"ACTTAAAGCAATATTAAAAAGTATAAAAA","2":"29.0","3":"1"},{"1":"AGTTTTAGCCTATTATTAATTGTTT","2":"25.0","3":"1"},{"1":"AGTTTTAGCCTATTGTTAATTGTTTGTATGATAAAATA","2":"38.0","3":"1"},{"1":"AGTTTTAGCCTATTGTTAATTGTTTGTATGGTAAAAT","2":"37.0","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"35.7","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"36.8","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.2","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"37.3","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATCAA","2":"40.0","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATGT","2":"39.0","3":"1"},{"1":"AGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"38.0","3":"1"},{"1":"ATAAAGAAATTTAAAAAGGGACTAAAAC","2":"28.0","3":"1"},{"1":"ATAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"1"},{"1":"ATATATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"40.0","3":"1"},{"1":"ATCTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"40.0","3":"1"},{"1":"ATCTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"37.0","3":"1"},{"1":"ATGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"1"},{"1":"ATTAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"40.0","3":"1"},{"1":"ATTTTACCAAAAAGAAATTTAAAAAGGGGCTAAAACA","2":"36.3","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAA","2":"25.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAA","2":"34.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"35.7","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"35.7","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"36.3","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"37.2","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAC","2":"38.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAG","2":"38.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAT","2":"38.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACATT","2":"39.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACGA","2":"38.0","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"36.3","3":"1"},{"1":"ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"36.3","3":"1"},{"1":"ATTTTACCATAAAGAAGTTTAAAAAGGGACTAAAACT","2":"37.0","3":"1"},{"1":"ATTTTACCATAATGAAATTTAAAAAGAGACTAAAAC","2":"36.0","3":"1"},{"1":"ATTTTACCATACAAACAATTAACAATAGGCTAAAACT","2":"37.0","3":"1"},{"1":"ATTTTACCATAGAGAAAATTAAAAAGAGACTAAAAC","2":"32.7","3":"1"},{"1":"ATTTTACCATATAAACAATTAACAATAGGCTAAAAC","2":"35.7","3":"1"},{"1":"ATTTTACCATGTAAACAATTAACAAGAGACTAAAACC","2":"37.0","3":"1"},{"1":"ATTTTACCATGTAAACAATTAACAATAGGCTAAAAC","2":"36.0","3":"1"},{"1":"ATTTTACCATGTAAACAATTAATAATA","2":"27.0","3":"1"},{"1":"ATTTTAGCCTATTATTAATTGTTTATATGGTAAAATA","2":"37.0","3":"1"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAA","2":"38.0","3":"1"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAAA","2":"39.0","3":"1"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAT","2":"38.0","3":"1"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATATA","2":"39.0","3":"1"},{"1":"ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTAAT","2":"40.0","3":"1"},{"1":"ATTTTATCATACAAACAATTAATAATAGGCTAAAACA","2":"37.0","3":"1"},{"1":"ATTTTATCATACAAACATTTAACAAGAGGCTAAAAC","2":"36.0","3":"1"},{"1":"ATTTTATCATATAAACAATTAATAATAGGCTAAAAC","2":"35.7","3":"1"},{"1":"ATTTTATCATATAAACAATTAATAATAGGCTAAAAC","2":"35.8","3":"1"},{"1":"ATTTTATCATATAAACAATTAATAATCTGTCTCTTA","2":"36.0","3":"1"},{"1":"ATTTTGCCATAAAGAAATTTAAAAAGGGACTAAAACA","2":"36.7","3":"1"},{"1":"CATACAAACATTTAACAAGAGGCTAAAAC","2":"29.0","3":"1"},{"1":"CATATAAACAATTAATAATAGGCTAAAA","2":"28.0","3":"1"},{"1":"CATTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"40.0","3":"1"},{"1":"CATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAA","2":"39.0","3":"1"},{"1":"CATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATTA","2":"39.0","3":"1"},{"1":"CCATAAAGAAATTTAAAAAGGGACTAAAA","2":"29.0","3":"1"},{"1":"CCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"30.0","3":"1"},{"1":"CCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"31.0","3":"1"},{"1":"CCCTTTTTAAATTTCTTTATGGTAAAATT","2":"29.0","3":"1"},{"1":"CCTATTATTAATTATTTATATGATAAAAT","2":"29.0","3":"1"},{"1":"CCTATTATTAATTGTTTATATGATAAAAT","2":"29.0","3":"1"},{"1":"CCTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAT","2":"39.0","3":"1"},{"1":"CCTTTTTAAATTTCTTTATGGTA","2":"23.0","3":"1"},{"1":"CGATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATTT","2":"40.0","3":"1"},{"1":"CTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAAAC","2":"40.0","3":"1"},{"1":"CTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACT","2":"39.0","3":"1"},{"1":"CTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATTT","2":"40.0","3":"1"},{"1":"GAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACAA","2":"40.0","3":"1"},{"1":"GACGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"40.0","3":"1"},{"1":"GAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"38.0","3":"1"},{"1":"GATGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"40.0","3":"1"},{"1":"GCATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"38.0","3":"1"},{"1":"GCTTTAGCCTATCATTAATTGTTTATATGATAAAAT","2":"36.0","3":"1"},{"1":"GGGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"1"},{"1":"GGTTTTAGTCTCTTGTTAATTGTTTACATGGTAAAAT","2":"37.0","3":"1"},{"1":"GTAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"1"},{"1":"GTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"1"},{"1":"GTTGAATGCCAAAATCGATAAGGTTGAT","2":"28.0","3":"1"},{"1":"GTTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"1"},{"1":"GTTTTAACCTATTATTAATTGTTTATATGGTAAAAT","2":"36.0","3":"1"},{"1":"GTTTTAGCCTATTATTAATTATTTATAT","2":"28.0","3":"1"},{"1":"GTTTTAGCCTATTATTAATTGTTTATAT","2":"28.0","3":"1"},{"1":"GTTTTAGCCTATTATTAATTGTTTATATGGTAAAAT","2":"35.8","3":"1"},{"1":"GTTTTAGCCTATTGTTAATTGTTTA","2":"25.0","3":"1"},{"1":"GTTTTAGCCTCTTGTTAAATGTTTGTATGATAAAATA","2":"37.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATG","2":"29.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGG","2":"30.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAA","2":"35.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"36.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"36.8","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAA","2":"38.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAGA","2":"39.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAGAT","2":"40.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATATA","2":"39.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATCA","2":"38.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATGT","2":"38.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTG","2":"38.0","3":"1"},{"1":"GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTT","2":"38.0","3":"1"},{"1":"TAAACAATTAACAATAGGCTAAAACA","2":"26.0","3":"1"},{"1":"TAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAAA","2":"39.0","3":"1"},{"1":"TAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATA","2":"39.0","3":"1"},{"1":"TAGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATA","2":"39.0","3":"1"},{"1":"TATAATCTCATAAGAAATTTAAAAAGGGACTAAAAC","2":"36.0","3":"1"},{"1":"TATATTTTACCATAAAGAAATTTAAAAAGGGACTAACCC","2":"39.0","3":"1"},{"1":"TATTTTACCATAAAGAAAATTAAAAAGGGACTAA","2":"35.7","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAA","2":"24.0","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"35.7","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACC","2":"38.0","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACGA","2":"39.0","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATA","2":"38.0","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATAAA","2":"40.0","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGACTAAAATAT","2":"39.0","3":"1"},{"1":"TATTTTACCATAAAGAAATTTAAAAAGGGGCTAAAATAA","2":"39.0","3":"1"},{"1":"TATTTTACCATATAAACAATTAATAATAGGCTAAAA","2":"36.0","3":"1"},{"1":"TATTTTACCATATAAACAATTAATAATAGGCTAAAAC","2":"37.0","3":"1"},{"1":"TATTTTAGCCCCTTTTTAAATTTCTTTATGATAAAAT","2":"37.0","3":"1"},{"1":"TATTTTATCATACAAACAATTAACAATAGGCTAAAAC","2":"37.0","3":"1"},{"1":"TATTTTATCATACAAACATTTAA","2":"23.0","3":"1"},{"1":"TATTTTATCATATAAACAATTAATAATAGGCTAAAACA","2":"38.0","3":"1"},{"1":"TCCATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"39.0","3":"1"},{"1":"TGATTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTC","2":"40.0","3":"1"},{"1":"TGCATTTTACCATAAAGAAATTTAAAAAGGGACTAA","2":"36.0","3":"1"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGCAAAAT","2":"36.7","3":"1"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGCAAAATT","2":"37.3","3":"1"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAA","2":"35.0","3":"1"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"35.0","3":"1"},{"1":"TGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTG","2":"39.0","3":"1"},{"1":"TTATTAATTGTTTATATGGTAAAAT","2":"25.0","3":"1"},{"1":"TTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAA","2":"37.0","3":"1"},{"1":"TTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"38.0","3":"1"},{"1":"TTATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"39.0","3":"1"},{"1":"TTGTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATG","2":"39.0","3":"1"},{"1":"TTTAATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"40.0","3":"1"},{"1":"TTTACCATAAAGAAATTTAAAAAGGGACTAAAA","2":"33.0","3":"1"},{"1":"TTTATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT","2":"39.0","3":"1"},{"1":"TTTTAAATTTCTTTATGGTAAAATA","2":"25.0","3":"1"},{"1":"TTTTACCATAAAGAAATTTAAAAAGGGACTAAAAAACCCT","2":"40.0","3":"1"},{"1":"TTTTACCATAAAGAAATTTAAAAAGGGACTAAAACTTTCA","2":"40.0","3":"1"},{"1":"TTTTACCATAAAGAAATTTAAAAAGGGACTAAAATAAACC","2":"40.0","3":"1"},{"1":"TTTTACCATATAAACAATTAATAATAGGCTAAAACTTTGC","2":"40.0","3":"1"},{"1":"TTTTAGCCTATTATTAATTGTTTA","2":"24.0","3":"1"},{"1":"TTTTAGCCTATTATTAATTGTTTACATGGTAAAAT","2":"35.0","3":"1"},{"1":"TTTTAGCCTATTATTAATTGTTTATATGATAAAATA","2":"36.0","3":"1"},{"1":"TTTTAGCCTATTGTTAATTGTTTGTATGATAAAATA","2":"36.0","3":"1"},{"1":"TTTTAGCCTCTTGTTAAATGTTTGTATGATAAAAT","2":"35.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTA","2":"26.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGG","2":"29.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGT","2":"30.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTA","2":"31.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAA","2":"32.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAA","2":"33.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"36.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATT","2":"37.0","3":"1"},{"1":"TTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATTC","2":"37.0","3":"1"},{"1":"TTTTATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC","2":"40.0","3":"1"},{"1":"TTTTTAGCCTATTATTAATTGTTTATATGATAAAATT","2":"37.0","3":"1"},{"1":"TTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAATAAT","2":"39.0","3":"1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>The top two consensus sequences occur evenly frequently and are each others’ reverse complement. The next two are variants that lack either the first or last nucleotide (both from the same strand, ‘AT-start’), at roughly equal frequencies. The fifth and sixth are one nucleotide longer (from either strand, with an A appended or T prepended = identical). The next sequence is actually different (<code>ATTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT</code>), still at 36 bp long. Then follow one with a single point mutation and some length variants. Some clustering (with <code>CDHIT-EST</code>) should probably be able to help find the actually different repeat sequences.</p>
<p>Most common repeat (in ‘AT-start’ orientation): <code>ATTTTACCATAAAGAAATTTAAAAAGGGACTAAAAC</code></p>
<p>For comparison, the ‘GT-start’ orientation (reverse complement): <code>GTTTTAGTCCCTTTTTAAATTTCTTTATGGTAAAAT</code></p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>